// @strudel/core/dist/index.mjs
import S from "fraction.js";
const Dt = "strudel.log";
let Oe = 1e3, Wt, $t;
function Gt(t, e = "cyclist") {
  process.env.NODE_ENV === "development" && console.error(t), E(`[${e}] error: ${t.message}`);
}
function E(t, e, n = {}) {
  let s = performance.now();
  Wt === t && s - $t < Oe || (Wt = t, $t = s, console.log(`%c${t}`, "background-color: black;color:white;border-radius:15px"), typeof document < "u" && typeof CustomEvent < "u" && document.dispatchEvent(
    new CustomEvent(Dt, {
      detail: {
        message: t,
        type: e,
        data: n
      }
    })
  ));
}
E.key = Dt;
const $a = (t) => /^[a-gA-G][#bs]*[0-9]$/.test(t), qt = (t) => /^[a-gA-G][#bsf]*-?[0-9]?$/.test(t), ze = (t) => {
  if (typeof t != "string")
    return [];
  const [e, n = "", s] = t.match(/^([a-gA-G])([#bsf]*)(-?[0-9]*)$/)?.slice(1) || [];
  return e ? [e, n, s ? Number(s) : void 0] : [];
}, Ee = { c: 0, d: 2, e: 4, f: 5, g: 7, a: 9, b: 11 }, Me = { "#": 1, b: -1, s: 1, f: -1 }, pt = (t, e = 3) => {
  const [n, s, r = e] = ze(t);
  if (!n)
    throw new Error('not a note: "' + t + '"');
  const o = Ee[n.toLowerCase()], c = s?.split("").reduce((u, a) => u + Me[a], 0) || 0;
  return (Number(r) + 1) * 12 + o + c;
}, nt = (t) => Math.pow(2, (t - 69) / 12) * 440, Pe = (t) => 12 * Math.log(t / 440) / Math.LN2 + 69, Ra = (t, e) => {
  if (typeof t != "object")
    throw new Error("valueToMidi: expected object value");
  let { freq: n, note: s } = t;
  if (typeof n == "number")
    return Pe(n);
  if (typeof s == "string")
    return pt(s);
  if (typeof s == "number")
    return s;
  if (!e)
    throw new Error("valueToMidi: expected freq or note to be set");
  return e;
}, La = (t, e) => (t - e) * 1e3, Je = (t) => nt(typeof t == "number" ? t : pt(t)), je = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"], Ia = (t) => {
  const e = Math.floor(t / 12) - 1;
  return je[t % 12] + e;
}, ft = (t, e) => (t % e + e) % e, Ne = (t) => t.reduce((e, n) => e + n) / t.length;
function We(t, e = 0) {
  return isNaN(Number(t)) ? (E(`"${t}" is not a number, falling back to ${e}`, "warning"), e) : t;
}
const Ha = (t, e) => ft(Math.round(We(t ?? 0, 0)), e), Va = (t) => {
  let { value: e, context: n } = t, s = e;
  if (typeof s == "object" && !Array.isArray(s) && (s = s.note || s.n || s.value, s === void 0))
    throw new Error(`cannot find a playable note for ${JSON.stringify(e)}`);
  if (typeof s == "number" && n.type !== "frequency")
    s = nt(t.value);
  else if (typeof s == "number" && n.type === "frequency")
    s = t.value;
  else if (typeof s != "string" || !qt(s))
    throw new Error("not a note: " + JSON.stringify(s));
  return s;
}, Fa = (t) => {
  let { value: e, context: n } = t;
  if (typeof e == "object")
    return e.freq ? e.freq : Je(e.note || e.n || e.value);
  if (typeof e == "number" && n.type !== "frequency")
    e = nt(t.value);
  else if (typeof e == "string" && qt(e))
    e = nt(pt(t.value));
  else if (typeof e != "number")
    throw new Error("not a note or frequency: " + e);
  return e;
}, $e = (t, e) => t.slice(e).concat(t.slice(0, e)), Re = (...t) => t.reduce(
  (e, n) => (...s) => e(n(...s)),
  (e) => e
), Da = (...t) => Re(...t.reverse()), it = (t) => t.filter((e) => e != null), D = (t) => [].concat(...t), et = (t) => t, Ga = (t, e) => t, ht = (t, e) => Array.from({ length: e - t + 1 }, (n, s) => s + t);
function w(t, e, n = t.length) {
  const s = function r(...o) {
    if (o.length >= n)
      return t.apply(this, o);
    {
      const c = function(...u) {
        return r.apply(this, o.concat(u));
      };
      return e && e(c, o), c;
    }
  };
  return e && e(s, []), s;
}
function Qt(t) {
  const e = Number(t);
  if (!isNaN(e))
    return e;
  if (qt(t))
    return pt(t);
  throw new Error(`cannot parse as numeral: "${t}"`);
}
function Ut(t, e) {
  return (...n) => t(...n.map(e));
}
function $(t) {
  return Ut(t, Qt);
}
function Le(t) {
  const e = Number(t);
  if (!isNaN(e))
    return e;
  const n = {
    pi: Math.PI,
    w: 1,
    h: 0.5,
    q: 0.25,
    e: 0.125,
    s: 0.0625,
    t: 1 / 3,
    f: 0.2,
    x: 1 / 6
  }[t];
  if (typeof n < "u")
    return n;
  throw new Error(`cannot parse as fractional: "${t}"`);
}
const Qa = (t) => Ut(t, Le), Kt = function(t, e) {
  return [e.slice(0, t), e.slice(t)];
}, At = (t, e, n) => e.map((s, r) => t(s, n[r])), Ie = function(t) {
  const e = [];
  for (let n = 0; n < t.length - 1; ++n)
    e.push([t[n], t[n + 1]]);
  return e;
}, He = (t, e, n) => Math.min(Math.max(t, e), n), Ve = ["Do", "Reb", "Re", "Mib", "Mi", "Fa", "Solb", "Sol", "Lab", "La", "Sib", "Si"], Fe = [
  "Sa",
  "Re",
  "Ga",
  "Ma",
  "Pa",
  "Dha",
  "Ni"
], De = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Hb", "H"], Ge = [
  "Ni",
  "Pab",
  "Pa",
  "Voub",
  "Vou",
  "Ga",
  "Dib",
  "Di",
  "Keb",
  "Ke",
  "Zob",
  "Zo"
], Qe = [
  "I",
  "Ro",
  "Ha",
  "Ni",
  "Ho",
  "He",
  "To"
], Ue = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"], Ua = (t, e = "letters") => {
  const s = (e === "solfeggio" ? Ve : e === "indian" ? Fe : e === "german" ? De : e === "byzantine" ? Ge : e === "japanese" ? Qe : Ue)[t % 12], r = Math.floor(t / 12) - 1;
  return s + r;
};
function Ka(t) {
  var e = {};
  return t.filter(function(n) {
    return e.hasOwn(n) ? !1 : e[n] = !0;
  });
}
function Xa(t) {
  return t.sort().filter(function(e, n, s) {
    return !n || e != s[n - 1];
  });
}
function Ke(t) {
  return t.sort((e, n) => e.compare(n)).filter(function(e, n, s) {
    return !n || e.ne(s[n - 1]);
  });
}
function Xe(t) {
  const e = new TextEncoder().encode(t);
  return btoa(String.fromCharCode(...e));
}
function Ye(t) {
  const e = new Uint8Array(
    atob(t).split("").map((s) => s.charCodeAt(0))
  );
  return new TextDecoder().decode(e);
}
function Ya(t) {
  return encodeURIComponent(Xe(t));
}
function Za(t) {
  return Ye(decodeURIComponent(t));
}
function Ze(t, e) {
  return Array.isArray(t) ? t.map(e) : Object.fromEntries(Object.entries(t).map(([n, s], r) => [n, e(s, n, r)]));
}
function Rt(t, e) {
  return t / e;
}
class tn {
  constructor({
    getTargetClockTime: e = en,
    weight: n = 16,
    offsetDelta: s = 5e-3,
    checkAfterTime: r = 2,
    resetAfterTime: o = 8
  }) {
    this.offsetTime, this.timeAtPrevOffsetSample, this.prevOffsetTimes = [], this.getTargetClockTime = e, this.weight = n, this.offsetDelta = s, this.checkAfterTime = r, this.resetAfterTime = o, this.reset = () => {
      this.prevOffsetTimes = [], this.offsetTime = null, this.timeAtPrevOffsetSample = null;
    };
  }
  calculateOffset(e) {
    const n = this.getTargetClockTime(), s = n - this.timeAtPrevOffsetSample, r = n - e;
    if (s > this.resetAfterTime && this.reset(), this.offsetTime == null && (this.offsetTime = r), this.prevOffsetTimes.push(r), this.prevOffsetTimes.length > this.weight && this.prevOffsetTimes.shift(), this.timeAtPrevOffsetSample == null || s > this.checkAfterTime) {
      this.timeAtPrevOffsetSample = n;
      const o = Ne(this.prevOffsetTimes);
      Math.abs(o - this.offsetTime) > this.offsetDelta && (this.offsetTime = o);
    }
    return this.offsetTime;
  }
  calculateTimestamp(e, n) {
    return this.calculateOffset(e) + n;
  }
}
function tl() {
  return performance.now() * 1e-3;
}
function en() {
  return Date.now() * 1e-3;
}
const nn = /* @__PURE__ */ new Map([
  ["control", "Control"],
  ["ctrl", "Control"],
  ["alt", "Alt"],
  ["shift", "Shift"],
  ["down", "ArrowDown"],
  ["up", "ArrowUp"],
  ["left", "ArrowLeft"],
  ["right", "ArrowRight"]
]);
let tt;
function sn() {
  if (tt == null) {
    if (typeof window > "u")
      return;
    tt = {}, window.addEventListener("keydown", (t) => {
      tt[t.key] = !0;
    }), window.addEventListener("keyup", (t) => {
      tt[t.key] = !1;
    });
  }
  return { ...tt };
}
function Xt(t, e = !1) {
  return typeof t == "object" ? e ? JSON.stringify(t).slice(1, -1).replaceAll('"', "").replaceAll(",", " ") : JSON.stringify(t) : t;
}
S.prototype.sam = function() {
  return this.floor();
};
S.prototype.nextSam = function() {
  return this.sam().add(1);
};
S.prototype.wholeCycle = function() {
  return new x(this.sam(), this.nextSam());
};
S.prototype.cyclePos = function() {
  return this.sub(this.sam());
};
S.prototype.lt = function(t) {
  return this.compare(t) < 0;
};
S.prototype.gt = function(t) {
  return this.compare(t) > 0;
};
S.prototype.lte = function(t) {
  return this.compare(t) <= 0;
};
S.prototype.gte = function(t) {
  return this.compare(t) >= 0;
};
S.prototype.eq = function(t) {
  return this.compare(t) == 0;
};
S.prototype.ne = function(t) {
  return this.compare(t) != 0;
};
S.prototype.max = function(t) {
  return this.gt(t) ? this : t;
};
S.prototype.maximum = function(...t) {
  return t = t.map((e) => new S(e)), t.reduce((e, n) => n.max(e), this);
};
S.prototype.min = function(t) {
  return this.lt(t) ? this : t;
};
S.prototype.mulmaybe = function(t) {
  return t !== void 0 ? this.mul(t) : void 0;
};
S.prototype.divmaybe = function(t) {
  return t !== void 0 ? this.div(t) : void 0;
};
S.prototype.addmaybe = function(t) {
  return t !== void 0 ? this.add(t) : void 0;
};
S.prototype.submaybe = function(t) {
  return t !== void 0 ? this.sub(t) : void 0;
};
S.prototype.show = function() {
  return this.s * this.n + "/" + this.d;
};
S.prototype.or = function(t) {
  return this.eq(0) ? t : this;
};
const h = (t) => S(t), rn = (...t) => {
  if (t = it(t), t.length !== 0)
    return t.reduce((e, n) => e.gcd(n), h(1));
}, U = (...t) => {
  if (t = it(t), t.length === 0)
    return;
  const e = t.pop();
  return t.reduce(
    (n, s) => n === void 0 || s === void 0 ? void 0 : n.lcm(s),
    e
  );
}, on = (t) => t instanceof S;
h._original = S;
class x {
  constructor(e, n) {
    this.begin = h(e), this.end = h(n);
  }
  get spanCycles() {
    const e = [];
    var n = this.begin;
    const s = this.end, r = s.sam();
    if (n.equals(s))
      return [new x(n, s)];
    for (; s.gt(n); ) {
      if (n.sam().equals(r)) {
        e.push(new x(n, this.end));
        break;
      }
      const o = n.nextSam();
      e.push(new x(n, o)), n = o;
    }
    return e;
  }
  get duration() {
    return this.end.sub(this.begin);
  }
  cycleArc() {
    const e = this.begin.cyclePos(), n = e.add(this.duration);
    return new x(e, n);
  }
  withTime(e) {
    return new x(e(this.begin), e(this.end));
  }
  withEnd(e) {
    return new x(this.begin, e(this.end));
  }
  withCycle(e) {
    const n = this.begin.sam(), s = n.add(e(this.begin.sub(n))), r = n.add(e(this.end.sub(n)));
    return new x(s, r);
  }
  intersection(e) {
    const n = this.begin.max(e.begin), s = this.end.min(e.end);
    if (!n.gt(s) && !(n.equals(s) && (n.equals(this.end) && this.begin.lt(this.end) || n.equals(e.end) && e.begin.lt(e.end))))
      return new x(n, s);
  }
  intersection_e(e) {
    const n = this.intersection(e);
    if (n == null)
      throw "TimeSpans do not intersect";
    return n;
  }
  midpoint() {
    return this.begin.add(this.duration.div(h(2)));
  }
  equals(e) {
    return this.begin.equals(e.begin) && this.end.equals(e.end);
  }
  show() {
    return this.begin.show() + " → " + this.end.show();
  }
}
class C {
  /*
        Event class, representing a value active during the timespan
        'part'. This might be a fragment of an event, in which case the
        timespan will be smaller than the 'whole' timespan, otherwise the
        two timespans will be the same. The 'part' must never extend outside of the
        'whole'. If the event represents a continuously changing value
        then the whole will be returned as None, in which case the given
        value will have been sampled from the point halfway between the
        start and end of the 'part' timespan.
        The context is to store a list of source code locations causing the event.
  
        The word 'Event' is more or less a reserved word in javascript, hence this
        class is named called 'Hap'.
        */
  constructor(e, n, s, r = {}, o = !1) {
    this.whole = e, this.part = n, this.value = s, this.context = r, this.stateful = o, o && console.assert(typeof this.value == "function", "Stateful values must be functions");
  }
  get duration() {
    let e;
    return typeof this.value?.duration == "number" ? e = h(this.value.duration) : e = this.whole.end.sub(this.whole.begin), typeof this.value?.clip == "number" ? e.mul(this.value.clip) : e;
  }
  get endClipped() {
    return this.whole.begin.add(this.duration);
  }
  isActive(e) {
    return this.whole.begin <= e && this.endClipped >= e;
  }
  isInPast(e) {
    return e > this.endClipped;
  }
  isInNearPast(e, n) {
    return n - e <= this.endClipped;
  }
  isInFuture(e) {
    return e < this.whole.begin;
  }
  isInNearFuture(e, n) {
    return n < this.whole.begin && n > this.whole.begin - e;
  }
  isWithinTime(e, n) {
    return this.whole.begin <= n && this.endClipped >= e;
  }
  wholeOrPart() {
    return this.whole ? this.whole : this.part;
  }
  withSpan(e) {
    const n = this.whole ? e(this.whole) : void 0;
    return new C(n, e(this.part), this.value, this.context);
  }
  withValue(e) {
    return new C(this.whole, this.part, e(this.value), this.context);
  }
  hasOnset() {
    return this.whole != null && this.whole.begin.equals(this.part.begin);
  }
  hasTag(e) {
    return this.context.tags?.includes(e);
  }
  resolveState(e) {
    if (this.stateful && this.hasOnset()) {
      console.log("stateful");
      const n = this.value, [s, r] = n(e);
      return [s, new C(this.whole, this.part, r, this.context, !1)];
    }
    return [e, this];
  }
  spanEquals(e) {
    return this.whole == null && e.whole == null || this.whole.equals(e.whole);
  }
  equals(e) {
    return this.spanEquals(e) && this.part.equals(e.part) && // TODO would == be better ??
    this.value === e.value;
  }
  show(e = !1) {
    const n = typeof this.value == "object" ? e ? JSON.stringify(this.value).slice(1, -1).replaceAll('"', "").replaceAll(",", " ") : JSON.stringify(this.value) : this.value;
    var s = "";
    if (this.whole == null)
      s = "~" + this.part.show;
    else {
      var r = this.whole.begin.equals(this.part.begin) && this.whole.end.equals(this.part.end);
      this.whole.begin.equals(this.part.begin) || (s = this.whole.begin.show() + " ⇜ "), r || (s += "("), s += this.part.show(), r || (s += ")"), this.whole.end.equals(this.part.end) || (s += " ⇝ " + this.whole.end.show());
    }
    return "[ " + s + " | " + n + " ]";
  }
  showWhole(e = !1) {
    return `${this.whole == null ? "~" : this.whole.show()}: ${Xt(this.value, e)}`;
  }
  combineContext(e) {
    const n = this;
    return { ...n.context, ...e.context, locations: (n.context.locations || []).concat(e.context.locations || []) };
  }
  setContext(e) {
    return new C(this.whole, this.part, this.value, e);
  }
  ensureObjectValue() {
    if (typeof this.value != "object")
      throw new Error(
        `expected hap.value to be an object, but got "${this.value}". Hint: append .note() or .s() to the end`,
        "error"
      );
  }
}
class st {
  constructor(e, n = {}) {
    this.span = e, this.controls = n;
  }
  // Returns new State with different span
  setSpan(e) {
    return new st(e, this.controls);
  }
  withSpan(e) {
    return this.setSpan(e(this.span));
  }
  // Returns new State with added controls.
  setControls(e) {
    return new st(this.span, { ...this.controls, ...e });
  }
}
function cn(t, e, n) {
  if (e?.value !== void 0 && Object.keys(e).length === 1)
    return E("[warn]: Can't do arithmetic on control pattern."), t;
  const s = Object.keys(t).filter((r) => Object.keys(e).includes(r));
  return Object.assign({}, t, e, Object.fromEntries(s.map((r) => [r, n(t[r], e[r])])));
}
w((t, e) => t * e);
w((t, e) => e.map(t));
function un(t, e = 60) {
  let n = 0, s = h(0), r = [""], o = "";
  for (; r[0].length < e; ) {
    const c = t.queryArc(n, n + 1), u = c.filter((m) => m.hasOnset()).map((m) => m.duration), a = rn(...u), p = a.inverse();
    r = r.map((m) => m + "|"), o += "|";
    for (let m = 0; m < p; m++) {
      const [y, _] = [s, s.add(a)], g = c.filter((q) => q.whole.begin.lte(y) && q.whole.end.gte(_)), k = g.length - r.length;
      k > 0 && (r = r.concat(Array(k).fill(o))), r = r.map((q, M) => {
        const P = g[M];
        if (P) {
          const T = P.whole.begin.eq(y) ? "" + P.value : "-";
          return q + T;
        }
        return q + ".";
      }), o += ".", s = s.add(a);
    }
    n++;
  }
  return r.join(`
`);
}
let _t, j = !0;
const el = function(t) {
  j = !!t;
}, nl = (t) => _t = t;
class f {
  /**
   * Create a pattern. As an end user, you will most likely not create a Pattern directly.
   *
   * @param {function} query - The function that maps a `State` to an array of `Hap`.
   * @noAutocomplete
   */
  constructor(e, n = void 0) {
    this.query = e, this._Pattern = !0, this._steps = n;
  }
  get _steps() {
    return this.__steps;
  }
  set _steps(e) {
    this.__steps = e === void 0 ? void 0 : h(e);
  }
  setSteps(e) {
    return this._steps = e, this;
  }
  withSteps(e) {
    return j ? new f(this.query, this._steps === void 0 ? void 0 : e(this._steps)) : this;
  }
  get hasSteps() {
    return this._steps !== void 0;
  }
  //////////////////////////////////////////////////////////////////////
  // Haskell-style functor, applicative and monadic operations
  /**
   * Returns a new pattern, with the function applied to the value of
   * each hap. It has the alias `fmap`.
   * @synonyms fmap
   * @param {Function} func to to apply to the value
   * @returns Pattern
   * @example
   * "0 1 2".withValue(v => v + 10).log()
   */
  withValue(e) {
    const n = new f((s) => this.query(s).map((r) => r.withValue(e)));
    return n._steps = this._steps, n;
  }
  // runs func on query state
  withState(e) {
    return new f((n) => this.query(e(n)));
  }
  /**
   * see `withValue`
   * @noAutocomplete
   */
  fmap(e) {
    return this.withValue(e);
  }
  /**
   * Assumes 'this' is a pattern of functions, and given a function to
   * resolve wholes, applies a given pattern of values to that
   * pattern of functions.
   * @param {Function} whole_func
   * @param {Function} func
   * @noAutocomplete
   * @returns Pattern
   */
  appWhole(e, n) {
    const s = this, r = function(o) {
      const c = s.query(o), u = n.query(o), a = function(p, m) {
        const y = p.part.intersection(m.part);
        if (y != null)
          return new C(
            e(p.whole, m.whole),
            y,
            p.value(m.value),
            m.combineContext(p)
          );
      };
      return D(
        c.map((p) => it(u.map((m) => a(p, m))))
      );
    };
    return new f(r);
  }
  /**
   * When this method is called on a pattern of functions, it matches its haps
   * with those in the given pattern of values.  A new pattern is returned, with
   * each matching value applied to the corresponding function.
   *
   * In this `_appBoth` variant, where timespans of the function and value haps
   * are not the same but do intersect, the resulting hap has a timespan of the
   * intersection. This applies to both the part and the whole timespan.
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appBoth(e) {
    const n = this, s = function(o, c) {
      if (!(o == null || c == null))
        return o.intersection_e(c);
    }, r = n.appWhole(s, e);
    return j && (r._steps = U(e._steps, n._steps)), r;
  }
  /**
   * As with `appBoth`, but the `whole` timespan is not the intersection,
   * but the timespan from the function of patterns that this method is called
   * on. In practice, this means that the pattern structure, including onsets,
   * are preserved from the pattern of functions (often referred to as the left
   * hand or inner pattern).
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appLeft(e) {
    const n = this, s = function(o) {
      const c = [];
      for (const u of n.query(o)) {
        const a = e.query(o.setSpan(u.wholeOrPart()));
        for (const p of a) {
          const m = u.whole, y = u.part.intersection(p.part);
          if (y) {
            const _ = u.value(p.value), g = p.combineContext(u), k = new C(m, y, _, g);
            c.push(k);
          }
        }
      }
      return c;
    }, r = new f(s);
    return r._steps = this._steps, r;
  }
  /**
   * As with `appLeft`, but `whole` timespans are instead taken from the
   * pattern of values, i.e. structure is preserved from the right hand/outer
   * pattern.
   * @param {Pattern} pat_val
   * @noAutocomplete
   * @returns Pattern
   */
  appRight(e) {
    const n = this, s = function(o) {
      const c = [];
      for (const u of e.query(o)) {
        const a = n.query(o.setSpan(u.wholeOrPart()));
        for (const p of a) {
          const m = u.whole, y = p.part.intersection(u.part);
          if (y) {
            const _ = p.value(u.value), g = u.combineContext(p), k = new C(m, y, _, g);
            c.push(k);
          }
        }
      }
      return c;
    }, r = new f(s);
    return r._steps = e._steps, r;
  }
  bindWhole(e, n) {
    const s = this, r = function(o) {
      const c = function(a, p) {
        return new C(
          e(a.whole, p.whole),
          p.part,
          p.value,
          Object.assign({}, a.context, p.context, {
            locations: (a.context.locations || []).concat(p.context.locations || [])
          })
        );
      }, u = function(a) {
        return n(a.value).query(o.setSpan(a.part)).map((p) => c(a, p));
      };
      return D(s.query(o).map((a) => u(a)));
    };
    return new f(r);
  }
  bind(e) {
    const n = function(s, r) {
      if (!(s == null || r == null))
        return s.intersection_e(r);
    };
    return this.bindWhole(n, e);
  }
  join() {
    return this.bind(et);
  }
  outerBind(e) {
    return this.bindWhole((n) => n, e).setSteps(this._steps);
  }
  outerJoin() {
    return this.outerBind(et);
  }
  innerBind(e) {
    return this.bindWhole((n, s) => s, e);
  }
  innerJoin() {
    return this.innerBind(et);
  }
  // Flatterns patterns of patterns, by retriggering/resetting inner patterns at onsets of outer pattern haps
  resetJoin(e = !1) {
    const n = this;
    return new f((s) => n.discreteOnly().query(s).map((r) => r.value.late(e ? r.whole.begin : r.whole.begin.cyclePos()).query(s).map(
      (o) => new C(
        // Supports continuous haps in the inner pattern
        o.whole ? o.whole.intersection(r.whole) : void 0,
        o.part.intersection(r.part),
        o.value
      ).setContext(r.combineContext(o))
    ).filter((o) => o.part)).flat());
  }
  restartJoin() {
    return this.resetJoin(!0);
  }
  // Like the other joins above, joins a pattern of patterns of values, into a flatter
  // pattern of values. In this case it takes whole cycles of the inner pattern to fit each event
  // in the outer pattern.
  squeezeJoin() {
    const e = this;
    function n(s) {
      const r = e.discreteOnly().query(s);
      function o(u) {
        const p = u.value._focusSpan(u.wholeOrPart()).query(s.setSpan(u.part));
        function m(y, _) {
          let g;
          if (_.whole && y.whole && (g = _.whole.intersection(y.whole), !g))
            return;
          const k = _.part.intersection(y.part);
          if (!k)
            return;
          const q = _.combineContext(y);
          return new C(g, k, _.value, q);
        }
        return p.map((y) => m(u, y));
      }
      return D(r.map(o)).filter((u) => u);
    }
    return new f(n);
  }
  squeezeBind(e) {
    return this.fmap(e).squeezeJoin();
  }
  polyJoin = function() {
    const e = this;
    return e.fmap((n) => n.extend(e._steps.div(n._steps))).outerJoin();
  };
  polyBind(e) {
    return this.fmap(e).polyJoin();
  }
  //////////////////////////////////////////////////////////////////////
  // Utility methods mainly for internal use
  /**
   * Query haps inside the given time span.
   *
   * @param {Fraction | number} begin from time
   * @param {Fraction | number} end to time
   * @returns Hap[]
   * @example
   * const pattern = sequence('a', ['b', 'c'])
   * const haps = pattern.queryArc(0, 1)
   * console.log(haps)
   * silence
   * @noAutocomplete
   */
  queryArc(e, n, s = {}) {
    try {
      return this.query(new st(new x(e, n), s));
    } catch (r) {
      return E(`[query]: ${r.message}`, "error"), [];
    }
  }
  /**
   * Returns a new pattern, with queries split at cycle boundaries. This makes
   * some calculations easier to express, as all haps are then constrained to
   * happen within a cycle.
   * @returns Pattern
   * @noAutocomplete
   */
  splitQueries() {
    const e = this, n = (s) => D(s.span.spanCycles.map((r) => e.query(s.setSpan(r))));
    return new f(n);
  }
  /**
   * Returns a new pattern, where the given function is applied to the query
   * timespan before passing it to the original pattern.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withQuerySpan(e) {
    return new f((n) => this.query(n.withSpan(e)));
  }
  withQuerySpanMaybe(e) {
    const n = this;
    return new f((s) => {
      const r = s.withSpan(e);
      return r.span ? n.query(r) : [];
    });
  }
  /**
   * As with `withQuerySpan`, but the function is applied to both the
   * begin and end time of the query timespan.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withQueryTime(e) {
    return new f((n) => this.query(n.withSpan((s) => s.withTime(e))));
  }
  /**
   * Similar to `withQuerySpan`, but the function is applied to the timespans
   * of all haps returned by pattern queries (both `part` timespans, and where
   * present, `whole` timespans).
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHapSpan(e) {
    return new f((n) => this.query(n).map((s) => s.withSpan(e)));
  }
  /**
   * As with `withHapSpan`, but the function is applied to both the
   * begin and end time of the hap timespans.
   * @param {Function} func the function to apply
   * @returns Pattern
   * @noAutocomplete
   */
  withHapTime(e) {
    return this.withHapSpan((n) => n.withTime(e));
  }
  /**
   * Returns a new pattern with the given function applied to the list of haps returned by every query.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHaps(e) {
    const n = new f((s) => e(this.query(s), s));
    return n._steps = this._steps, n;
  }
  /**
   * As with `withHaps`, but applies the function to every hap, rather than every list of haps.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withHap(e) {
    return this.withHaps((n) => n.map(e));
  }
  /**
   * Returns a new pattern with the context field set to every hap set to the given value.
   * @param {*} context
   * @returns Pattern
   * @noAutocomplete
   */
  setContext(e) {
    return this.withHap((n) => n.setContext(e));
  }
  /**
   * Returns a new pattern with the given function applied to the context field of every hap.
   * @param {Function} func
   * @returns Pattern
   * @noAutocomplete
   */
  withContext(e) {
    const n = this.withHap((s) => s.setContext(e(s.context)));
    return this.__pure !== void 0 && (n.__pure = this.__pure, n.__pure_loc = this.__pure_loc), n;
  }
  /**
   * Returns a new pattern with the context field of every hap set to an empty object.
   * @returns Pattern
   * @noAutocomplete
   */
  stripContext() {
    return this.withHap((e) => e.setContext({}));
  }
  /**
   * Returns a new pattern with the given location information added to the
   * context of every hap.
   * @param {Number} start start offset
   * @param {Number} end end offset
   * @returns Pattern
   * @noAutocomplete
   */
  withLoc(e, n) {
    const s = {
      start: e,
      end: n
    }, r = this.withContext((o) => {
      const c = (o.locations || []).concat([s]);
      return { ...o, locations: c };
    });
    return this.__pure && (r.__pure = this.__pure, r.__pure_loc = s), r;
  }
  /**
   * Returns a new Pattern, which only returns haps that meet the given test.
   * @param {Function} hap_test - a function which returns false for haps to be removed from the pattern
   * @returns Pattern
   * @noAutocomplete
   */
  filterHaps(e) {
    return new f((n) => this.query(n).filter(e));
  }
  /**
   * As with `filterHaps`, but the function is applied to values
   * inside haps.
   * @param {Function} value_test
   * @returns Pattern
   * @noAutocomplete
   */
  filterValues(e) {
    return new f((n) => this.query(n).filter((s) => e(s.value))).setSteps(this._steps);
  }
  /**
   * Returns a new pattern, with haps containing undefined values removed from
   * query results.
   * @returns Pattern
   * @noAutocomplete
   */
  removeUndefineds() {
    return this.filterValues((e) => e != null);
  }
  /**
   * Returns a new pattern, with all haps without onsets filtered out. A hap
   * with an onset is one with a `whole` timespan that begins at the same time
   * as its `part` timespan.
   * @returns Pattern
   * @noAutocomplete
   */
  onsetsOnly() {
    return this.filterHaps((e) => e.hasOnset());
  }
  /**
   * Returns a new pattern, with 'continuous' haps (those without 'whole'
   * timespans) removed from query results.
   * @returns Pattern
   * @noAutocomplete
   */
  discreteOnly() {
    return this.filterHaps((e) => e.whole);
  }
  /**
   * Combines adjacent haps with the same value and whole.  Only
   * intended for use in tests.
   * @noAutocomplete
   */
  defragmentHaps() {
    return this.discreteOnly().withHaps((n) => {
      const s = [];
      for (var r = 0; r < n.length; ++r) {
        for (var o = !0, c = n[r]; o; ) {
          const p = JSON.stringify(n[r].value);
          for (var u = !1, a = r + 1; a < n.length; a++) {
            const m = n[a];
            if (c.whole.equals(m.whole)) {
              if (c.part.begin.eq(m.part.end)) {
                if (p === JSON.stringify(m.value)) {
                  c = new C(c.whole, new x(m.part.begin, c.part.end), c.value), n.splice(a, 1), u = !0;
                  break;
                }
              } else if (m.part.begin.eq(c.part.end) && p == JSON.stringify(m.value)) {
                c = new C(c.whole, new x(c.part.begin, m.part.end), c.value), n.splice(a, 1), u = !0;
                break;
              }
            }
          }
          o = u;
        }
        s.push(c);
      }
      return s;
    });
  }
  /**
   * Queries the pattern for the first cycle, returning Haps. Mainly of use when
   * debugging a pattern.
   * @param {Boolean} with_context - set to true, otherwise the context field
   * will be stripped from the resulting haps.
   * @returns [Hap]
   * @noAutocomplete
   */
  firstCycle(e = !1) {
    var n = this;
    return e || (n = n.stripContext()), n.query(new st(new x(h(0), h(1))));
  }
  /**
   * Accessor for a list of values returned by querying the first cycle.
   * @noAutocomplete
   */
  get firstCycleValues() {
    return this.firstCycle().map((e) => e.value);
  }
  /**
   * More human-readable version of the `firstCycleValues` accessor.
   * @noAutocomplete
   */
  get showFirstCycle() {
    return this.firstCycle().map(
      (e) => `${e.value}: ${e.whole.begin.toFraction()} - ${e.whole.end.toFraction()}`
    );
  }
  /**
   * Returns a new pattern, which returns haps sorted in temporal order. Mainly
   * of use when comparing two patterns for equality, in tests.
   * @returns Pattern
   * @noAutocomplete
   */
  sortHapsByPart() {
    return this.withHaps(
      (e) => e.sort(
        (n, s) => n.part.begin.sub(s.part.begin).or(n.part.end.sub(s.part.end)).or(n.whole.begin.sub(s.whole.begin).or(n.whole.end.sub(s.whole.end)))
      )
    );
  }
  asNumber() {
    return this.fmap(Qt);
  }
  //////////////////////////////////////////////////////////////////////
  // Operators - see 'make composers' later..
  _opIn(e, n) {
    return this.fmap(n).appLeft(d(e));
  }
  _opOut(e, n) {
    return this.fmap(n).appRight(d(e));
  }
  _opMix(e, n) {
    return this.fmap(n).appBoth(d(e));
  }
  _opSqueeze(e, n) {
    const s = d(e);
    return this.fmap((r) => s.fmap((o) => n(r)(o))).squeezeJoin();
  }
  _opSqueezeOut(e, n) {
    const s = this;
    return d(e).fmap((o) => s.fmap((c) => n(c)(o))).squeezeJoin();
  }
  _opReset(e, n) {
    return d(e).fmap((r) => this.fmap((o) => n(o)(r))).resetJoin();
  }
  _opRestart(e, n) {
    return d(e).fmap((r) => this.fmap((o) => n(o)(r))).restartJoin();
  }
  _opPoly(e, n) {
    const s = d(e);
    return this.fmap((r) => s.fmap((o) => n(o)(r))).polyJoin();
  }
  //////////////////////////////////////////////////////////////////////
  // End-user methods.
  // Those beginning with an underscore (_) are 'patternified',
  // i.e. versions are created without the underscore, that are
  // magically transformed to accept patterns for all their arguments.
  //////////////////////////////////////////////////////////////////////
  // Methods without corresponding toplevel functions
  /**
   * Layers the result of the given function(s). Like `superimpose`, but without the original pattern:
   * @name layer
   * @memberof Pattern
   * @synonyms apply
   * @returns Pattern
   * @example
   * "<0 2 4 6 ~ 4 ~ 2 0!3 ~!5>*8"
   *   .layer(x=>x.add("0,2"))
   *   .scale('C minor').note()
   */
  layer(...e) {
    return B(...e.map((n) => n(this)));
  }
  /**
   * Superimposes the result of the given function(s) on top of the original pattern:
   * @name superimpose
   * @memberof Pattern
   * @returns Pattern
   * @example
   * "<0 2 4 6 ~ 4 ~ 2 0!3 ~!5>*8"
   *   .superimpose(x=>x.add(2))
   *   .scale('C minor').note()
   */
  superimpose(...e) {
    return this.stack(...e.map((n) => n(this)));
  }
  //////////////////////////////////////////////////////////////////////
  // Multi-pattern functions
  stack(...e) {
    return B(this, ...e);
  }
  sequence(...e) {
    return G(this, ...e);
  }
  seq(...e) {
    return G(this, ...e);
  }
  cat(...e) {
    return at(this, ...e);
  }
  fastcat(...e) {
    return W(this, ...e);
  }
  slowcat(...e) {
    return K(this, ...e);
  }
  //////////////////////////////////////////////////////////////////////
  // Context methods - ones that deal with metadata
  onTrigger(e, n = !0) {
    return this.withHap(
      (s) => s.setContext({
        ...s.context,
        onTrigger: (...r) => {
          s.context.onTrigger?.(...r), e(...r);
        },
        // if dominantTrigger is set to true, the default output (webaudio) will be disabled
        // when using multiple triggers, you cannot flip this flag to false again!
        // example: x.csound('CooLSynth').log() as well as x.log().csound('CooLSynth') should work the same
        dominantTrigger: s.context.dominantTrigger || n
      })
    );
  }
  /**
   * Writes the content of the current event to the console (visible in the side menu).
   * @name log
   * @memberof Pattern
   * @example
   * s("bd sd").log()
   */
  log(e = (s) => `[hap] ${s.showWhole(!0)}`, n = (s) => ({ hap: s })) {
    return this.onTrigger((...s) => {
      E(e(...s), void 0, n(...s));
    }, !1);
  }
  /**
   * A simplified version of `log` which writes all "values" (various configurable parameters)
   * within the event to the console (visible in the side menu).
   * @name logValues
   * @memberof Pattern
   * @example
   * s("bd sd").gain("0.25 0.5 1").n("2 1 0").logValues()
   */
  logValues(e = (n) => `[hap] ${Xt(n, !0)}`) {
    return this.log((n) => e(n.value));
  }
  //////////////////////////////////////////////////////////////////////
  // Visualisation
  drawLine() {
    return console.log(un(this)), this;
  }
  //////////////////////////////////////////////////////////////////////
  // methods relating to breaking patterns into subcycles
  // Breaks a pattern into a pattern of patterns, according to the structure of the given binary pattern.
  unjoin(e, n = et) {
    return e.withHap(
      (s) => s.withValue((r) => r ? n(this.ribbon(s.whole.begin, s.whole.duration)) : this)
    );
  }
  /**
   * Breaks a pattern into pieces according to the structure of a given pattern.
   * True values in the given pattern cause the corresponding subcycle of the
   * source pattern to be looped, and for an (optional) given function to be
   * applied. False values result in the corresponding part of the source pattern
   * to be played unchanged.
   * @name into
   * @memberof Pattern
   * @example
   * sound("bd sd ht lt").into("1 0", hurry(2))
   */
  into(e, n) {
    return this.unjoin(e, n).innerJoin();
  }
}
function an(t, e) {
  let n = [];
  return e.forEach((s) => {
    const r = n.findIndex(([o]) => t(s, o));
    r === -1 ? n.push([s]) : n[r].push(s);
  }), n;
}
const ln = (t, e) => t.spanEquals(e);
f.prototype.collect = function() {
  return this.withHaps(
    (t) => an(ln, t).map((e) => new C(e[0].whole, e[0].part, e, {}))
  );
};
const sl = l("arpWith", (t, e) => e.collect().fmap((n) => d(t(n))).innerJoin().withHap((n) => new C(n.whole, n.part, n.value.value, n.combineContext(n.value)))), rl = l(
  "arp",
  (t, e) => e.arpWith((n) => d(t).fmap((s) => n[s % n.length])),
  !1
);
function ut(t) {
  return !Array.isArray(t) && typeof t == "object" && !on(t);
}
function pn(t, e, n) {
  return ut(t) || ut(e) ? (ut(t) || (t = { value: t }), ut(e) || (e = { value: e }), cn(t, e, n)) : n(t, e);
}
(function() {
  const t = {
    set: [(n, s) => s],
    keep: [(n) => n],
    keepif: [(n, s) => s ? n : void 0],
    // numerical functions
    /**
     *
     * Assumes a pattern of numbers. Adds the given number to each item in the pattern.
     * @name add
     * @memberof Pattern
     * @example
     * // Here, the triad 0, 2, 4 is shifted by different amounts
     * n("0 2 4".add("<0 3 4 0>")).scale("C:major")
     * // Without add, the equivalent would be:
     * // n("<[0 2 4] [3 5 7] [4 6 8] [0 2 4]>").scale("C:major")
     * @example
     * // You can also use add with notes:
     * note("c3 e3 g3".add("<0 5 7 0>"))
     * // Behind the scenes, the notes are converted to midi numbers:
     * // note("48 52 55".add("<0 5 7 0>"))
     */
    add: [$((n, s) => n + s)],
    // support string concatenation
    /**
     *
     * Like add, but the given numbers are subtracted.
     * @name sub
     * @memberof Pattern
     * @example
     * n("0 2 4".sub("<0 1 2 3>")).scale("C4:minor")
     * // See add for more information.
     */
    sub: [$((n, s) => n - s)],
    /**
     *
     * Multiplies each number by the given factor.
     * @name mul
     * @memberof Pattern
     * @example
     * "<1 1.5 [1.66, <2 2.33>]>*4".mul(150).freq()
     */
    mul: [$((n, s) => n * s)],
    /**
     *
     * Divides each number by the given factor.
     * @name div
     * @memberof Pattern
     */
    div: [$((n, s) => n / s)],
    mod: [$(ft)],
    pow: [$(Math.pow)],
    log2: [$(Math.log2)],
    band: [$((n, s) => n & s)],
    bor: [$((n, s) => n | s)],
    bxor: [$((n, s) => n ^ s)],
    blshift: [$((n, s) => n << s)],
    brshift: [$((n, s) => n >> s)],
    // TODO - force numerical comparison if both look like numbers?
    lt: [(n, s) => n < s],
    gt: [(n, s) => n > s],
    lte: [(n, s) => n <= s],
    gte: [(n, s) => n >= s],
    eq: [(n, s) => n == s],
    eqt: [(n, s) => n === s],
    ne: [(n, s) => n != s],
    net: [(n, s) => n !== s],
    and: [(n, s) => n && s],
    or: [(n, s) => n || s],
    //  bitwise ops
    func: [(n, s) => s(n)]
  }, e = ["In", "Out", "Mix", "Squeeze", "SqueezeOut", "Reset", "Restart", "Poly"];
  for (const [n, [s, r]] of Object.entries(t)) {
    f.prototype["_" + n] = function(o) {
      return this.fmap((c) => s(c, o));
    }, Object.defineProperty(f.prototype, n, {
      // a getter that returns a function, so 'pat' can be
      // accessed by closures that are methods of that function..
      get: function() {
        const o = this, c = (...u) => o[n].in(...u);
        for (const u of e)
          c[u.toLowerCase()] = function(...a) {
            var p = o;
            a = G(a), r && (p = r(p), a = r(a));
            var m;
            return n === "keepif" ? (m = p["_op" + u](a, (y) => (_) => s(y, _)), m = m.removeUndefineds()) : m = p["_op" + u](a, (y) => (_) => pn(y, _, s)), m;
          };
        return c.squeezein = c.squeeze, c;
      }
    });
    for (const o of e)
      f.prototype[o.toLowerCase()] = function(...c) {
        return this.set[o.toLowerCase()](c);
      };
  }
  f.prototype.struct = function(...n) {
    return this.keepif.out(...n);
  }, f.prototype.structAll = function(...n) {
    return this.keep.out(...n);
  }, f.prototype.mask = function(...n) {
    return this.keepif.in(...n);
  }, f.prototype.maskAll = function(...n) {
    return this.keep.in(...n);
  }, f.prototype.reset = function(...n) {
    return this.keepif.reset(...n);
  }, f.prototype.resetAll = function(...n) {
    return this.keep.reset(...n);
  }, f.prototype.restart = function(...n) {
    return this.keepif.restart(...n);
  }, f.prototype.restartAll = function(...n) {
    return this.keep.restart(...n);
  };
})();
const ol = B, il = B, cl = xt, ct = (t) => new f(() => [], t), v = ct(1), R = ct(0);
function z(t) {
  function e(s) {
    return s.span.spanCycles.map((r) => new C(h(r.begin).wholeCycle(), r, t));
  }
  const n = new f(e, 1);
  return n.__pure = t, n;
}
function Yt(t) {
  return t instanceof f || t?._Pattern;
}
function d(t) {
  return Yt(t) ? t : _t && typeof t == "string" ? _t(t) : z(t);
}
function fn(t) {
  let e = z([]);
  for (const n of t)
    e = e.bind((s) => n.fmap((r) => s.concat([r])));
  return e;
}
function B(...t) {
  t = t.map((s) => Array.isArray(s) ? G(...s) : d(s));
  const e = (s) => D(t.map((r) => r.query(s))), n = new f(e);
  return j && (n._steps = U(...t.map((s) => s._steps))), n;
}
function St(t, e) {
  if (e = e.map((o) => Array.isArray(o) ? G(...o) : d(o)), e.length === 0)
    return v;
  if (e.length === 1)
    return e[0];
  const [n, ...s] = e.map((o) => o._steps), r = j ? n.maximum(...s) : void 0;
  return B(...t(r, e));
}
function hn(...t) {
  return St(
    (e, n) => n.map((s) => s._steps.eq(e) ? s : N(s, ct(e.sub(s._steps)))),
    t
  );
}
function dn(...t) {
  return St(
    (e, n) => n.map((s) => s._steps.eq(e) ? s : N(ct(e.sub(s._steps)), s)),
    t
  );
}
function mn(...t) {
  return St(
    (e, n) => n.map((s) => {
      if (s._steps.eq(e))
        return s;
      const r = ct(e.sub(s._steps).div(2));
      return N(r, s, r);
    }),
    t
  );
}
function ul(t, ...e) {
  const [n, ...s] = e.map((c) => c._steps), r = n.maximum(...s), o = {
    centre: mn,
    left: hn,
    right: dn,
    expand: B,
    repeat: (...c) => xt(...c).steps(r)
  };
  return t.inhabit(o).fmap((c) => c(...e)).innerJoin().setSteps(r);
}
function K(...t) {
  if (t = t.map((s) => Array.isArray(s) ? W(...s) : d(s)), t.length == 1)
    return t[0];
  const e = function(s) {
    const r = s.span, o = ft(r.begin.sam(), t.length), c = t[o];
    if (!c)
      return [];
    const u = r.begin.floor().sub(r.begin.div(t.length).floor());
    return c.withHapTime((a) => a.add(u)).query(s.setSpan(r.withTime((a) => a.sub(u))));
  }, n = j ? U(...t.map((s) => s._steps)) : void 0;
  return new f(e).splitQueries().setSteps(n);
}
function Zt(...t) {
  t = t.map(d);
  const e = function(n) {
    const s = Math.floor(n.span.begin) % t.length;
    return t[s]?.query(n) || [];
  };
  return new f(e).splitQueries();
}
function at(...t) {
  return K(...t);
}
function al(...t) {
  const e = t.reduce((n, [s]) => n + s, 0);
  return t = t.map(([n, s]) => [n, s.fast(n)]), N(...t).slow(e);
}
function ll(...t) {
  let e = h(0);
  for (let n of t)
    n.length == 2 && n.unshift(e), e = n[1];
  return B(
    ...t.map(
      ([n, s, r]) => z(d(r)).compress(h(n).div(e), h(s).div(e))
    )
  ).slow(e).innerJoin();
}
function W(...t) {
  let e = K(...t);
  return t.length > 1 && (e = e._fast(t.length), e._steps = t.length), t.length == 1 && t[0].__steps_source && (t._steps = t[0]._steps), e;
}
function G(...t) {
  return W(...t);
}
function yn(...t) {
  return W(...t);
}
function vt(t) {
  return Array.isArray(t) ? t.length == 0 ? [v, 0] : t.length == 1 ? vt(t[0]) : [W(...t.map((e) => vt(e)[0])), t.length] : [d(t), 1];
}
const pl = w((t, e) => d(e).mask(t)), fl = w((t, e) => d(e).struct(t)), hl = w((t, e) => d(e).superimpose(...t)), dl = w((t, e) => d(e).withValue(t)), ml = w((t, e) => d(e).bind(t)), yl = w((t, e) => d(e).innerBind(t)), wl = w((t, e) => d(e).outerBind(t)), gl = w((t, e) => d(e).squeezeBind(t)), bl = w((t, e) => d(e).stepBind(t)), _l = w((t, e) => d(e).polyBind(t)), vl = w((t, e) => d(e).set(t)), kl = w((t, e) => d(e).keep(t)), ql = w((t, e) => d(e).keepif(t)), Al = w((t, e) => d(e).add(t)), Sl = w((t, e) => d(e).sub(t)), Tl = w((t, e) => d(e).mul(t)), Cl = w((t, e) => d(e).div(t)), xl = w((t, e) => d(e).mod(t)), Bl = w((t, e) => d(e).pow(t)), Ol = w((t, e) => d(e).band(t)), zl = w((t, e) => d(e).bor(t)), El = w((t, e) => d(e).bxor(t)), Ml = w((t, e) => d(e).blshift(t)), Pl = w((t, e) => d(e).brshift(t)), Jl = w((t, e) => d(e).lt(t)), jl = w((t, e) => d(e).gt(t)), Nl = w((t, e) => d(e).lte(t)), Wl = w((t, e) => d(e).gte(t)), $l = w((t, e) => d(e).eq(t)), Rl = w((t, e) => d(e).eqt(t)), Ll = w((t, e) => d(e).ne(t)), Il = w((t, e) => d(e).net(t)), Hl = w((t, e) => d(e).and(t)), Vl = w((t, e) => d(e).or(t)), Fl = w((t, e) => d(e).func(t));
function l(t, e, n = !0, s = !1, r = (o) => o.innerJoin()) {
  if (Array.isArray(t)) {
    const u = {};
    for (const a of t)
      u[a] = l(a, e, n, s, r);
    return u;
  }
  const o = e.length;
  var c;
  return n ? c = function(...u) {
    u = u.map(d);
    const a = u[u.length - 1];
    let p;
    if (o === 1)
      p = e(a);
    else {
      const m = u.slice(0, -1);
      if (m.every((y) => y.__pure != null)) {
        const y = m.map((g) => g.__pure), _ = m.filter((g) => g.__pure_loc).map((g) => g.__pure_loc);
        p = e(...y, a), p = p.withContext((g) => {
          const k = (g.locations || []).concat(_);
          return { ...g, locations: k };
        });
      } else {
        const [y, ..._] = m;
        let g = (...k) => e(...k, a);
        g = w(g, null, o - 1), p = r(_.reduce((k, q) => k.appLeft(q), y.fmap(g)));
      }
    }
    return s && (p._steps = a._steps), p;
  } : c = function(...u) {
    u = u.map(d);
    const a = e(...u);
    return s && (a._steps = u[u.length - 1]._steps), a;
  }, f.prototype[t] = function(...u) {
    if (o === 2 && u.length !== 1)
      u = [G(...u)];
    else if (o !== u.length + 1)
      throw new Error(`.${t}() expects ${o - 1} inputs but got ${u.length}.`);
    return u = u.map(d), c(...u, this);
  }, o > 1 && (f.prototype["_" + t] = function(...u) {
    const a = e(...u, this);
    return s && a.setSteps(this._steps), a;
  }), w(c, null, o);
}
function X(t, e, n = !0, s = !1, r = (o) => o.stepJoin()) {
  return l(t, e, n, s, r);
}
const Dl = l("round", function(t) {
  return t.asNumber().fmap((e) => Math.round(e));
}), Gl = l("floor", function(t) {
  return t.asNumber().fmap((e) => Math.floor(e));
}), Ql = l("ceil", function(t) {
  return t.asNumber().fmap((e) => Math.ceil(e));
}), Ul = l("toBipolar", function(t) {
  return t.fmap((e) => e * 2 - 1);
}), Kl = l("fromBipolar", function(t) {
  return t.fmap((e) => (e + 1) / 2);
}), Xl = l("range", function(t, e, n) {
  return n.mul(e - t).add(t);
}), Yl = l("rangex", function(t, e, n) {
  return n._range(Math.log(t), Math.log(e)).fmap(Math.exp);
}), Zl = l("range2", function(t, e, n) {
  return n.fromBipolar()._range(t, e);
}), tp = l(
  "ratio",
  (t) => t.fmap((e) => Array.isArray(e) ? e.slice(1).reduce((n, s) => n / s, e[0]) : e)
), ep = l("compress", function(t, e, n) {
  return t = h(t), e = h(e), t.gt(e) || t.gt(1) || e.gt(1) || t.lt(0) || e.lt(0) ? v : n._fastGap(h(1).div(e.sub(t)))._late(t);
}), { compressSpan: np, compressspan: sp } = l(["compressSpan", "compressspan"], function(t, e) {
  return e._compress(t.begin, t.end);
}), { fastGap: rp, fastgap: op } = l(["fastGap", "fastgap"], function(t, e) {
  const n = function(r) {
    const o = r.begin.sam(), c = r.begin.sub(o).mul(t).min(1), u = r.end.sub(o).mul(t).min(1);
    if (!(c >= 1))
      return new x(o.add(c), o.add(u));
  }, s = function(r) {
    const o = r.part.begin, c = r.part.end, u = o.sam(), a = o.sub(u).div(t).min(1), p = c.sub(u).div(t).min(1), m = new x(u.add(a), u.add(p)), y = r.whole ? new x(
      m.begin.sub(o.sub(r.whole.begin).div(t)),
      m.end.add(r.whole.end.sub(c).div(t))
    ) : void 0;
    return new C(y, m, r.value, r.context);
  };
  return e.withQuerySpanMaybe(n).withHap(s).splitQueries();
}), ip = l("focus", function(t, e, n) {
  return t = h(t), e = h(e), n._early(t.sam())._fast(h(1).div(e.sub(t)))._late(t);
}), { focusSpan: cp, focusspan: up } = l(["focusSpan", "focusspan"], function(t, e) {
  return e._focus(t.begin, t.end);
}), ap = l("ply", function(t, e) {
  const n = e.fmap((s) => z(s)._fast(t)).squeezeJoin();
  return j && (n._steps = h(t).mulmaybe(e._steps)), n;
}), { fast: lp, density: pp } = l(
  ["fast", "density"],
  function(t, e) {
    return t === 0 ? v : (t = h(t), e.withQueryTime((s) => s.mul(t)).withHapTime((s) => s.div(t)).setSteps(e._steps));
  },
  !0,
  !0
), fp = l("hurry", function(t, e) {
  return e._fast(t).mul(z({ speed: t }));
}), { slow: hp, sparsity: dp } = l(["slow", "sparsity"], function(t, e) {
  return t === 0 ? v : e._fast(h(1).div(t));
}), mp = l("inside", function(t, e, n) {
  return e(n._slow(t))._fast(t);
}), yp = l("outside", function(t, e, n) {
  return e(n._fast(t))._slow(t);
}), wp = l("lastOf", function(t, e, n) {
  const s = Array(t - 1).fill(n);
  return s.push(e(n)), Zt(...s);
}), { firstOf: gp, every: bp } = l(["firstOf", "every"], function(t, e, n) {
  const s = Array(t - 1).fill(n);
  return s.unshift(e(n)), Zt(...s);
}), _p = l("apply", function(t, e) {
  return t(e);
}), vp = l("cpm", function(t, e) {
  return e._fast(t / 60 / 1);
}), kp = l(
  "early",
  function(t, e) {
    return t = h(t), e.withQueryTime((n) => n.add(t)).withHapTime((n) => n.sub(t));
  },
  !0,
  !0
), wn = l(
  "late",
  function(t, e) {
    return t = h(t), e._early(h(0).sub(t));
  },
  !0,
  !0
), qp = l("zoom", function(t, e, n) {
  if (e = h(e), t = h(t), t.gte(e))
    return R;
  const s = e.sub(t), r = j ? n._steps?.mulmaybe(s) : void 0;
  return n.withQuerySpan((o) => o.withCycle((c) => c.mul(s).add(t))).withHapSpan((o) => o.withCycle((c) => c.sub(t).div(s))).splitQueries().setSteps(r);
}), { zoomArc: Ap, zoomarc: Sp } = l(["zoomArc", "zoomarc"], function(t, e) {
  return e.zoom(t.begin, t.end);
}), Tp = l(
  "bite",
  (t, e, n) => e.fmap((s) => (r) => {
    const o = h(s).div(r).mod(1), c = o.add(h(1).div(r));
    return n.zoom(o, c);
  }).appLeft(t).squeezeJoin(),
  !1
), Cp = l(
  "linger",
  function(t, e) {
    return t == 0 ? v : t < 0 ? e._zoom(t.add(1), 1)._slow(t) : e._zoom(0, t)._slow(t);
  },
  !0,
  !0
), { segment: xp, seg: Bp } = l(["segment", "seg"], function(t, e) {
  return e.struct(z(!0)._fast(t)).setSteps(t);
}), Op = l("swingBy", (t, e, n) => n.inside(e, wn(yn(0, t / 2)))), zp = l("swing", (t, e) => e.swingBy(1 / 3, t)), { invert: Ep, inv: Mp } = l(
  ["invert", "inv"],
  function(t) {
    return t.fmap((e) => !e);
  },
  !0,
  !0
), Pp = l("when", function(t, e, n) {
  return t ? e(n) : n;
}), Jp = l("off", function(t, e, n) {
  return B(n, e(n.late(t)));
}), jp = l("brak", function(t) {
  return t.when(K(!1, !0), (e) => W(e, v)._late(0.25));
}), gn = l(
  "rev",
  function(t) {
    const e = function(n) {
      const s = n.span, r = s.begin.sam(), o = s.begin.nextSam(), c = function(a) {
        const p = a.withTime((y) => r.add(o.sub(y))), m = p.begin;
        return p.begin = p.end, p.end = m, p;
      };
      return t.query(n.setSpan(c(s))).map((a) => a.withSpan(c));
    };
    return new f(e).splitQueries();
  },
  !1,
  !0
), Np = l("pressBy", function(t, e) {
  return e.fmap((n) => z(n).compress(t, 1)).squeezeJoin();
}), Wp = l("press", function(t) {
  return t._pressBy(0.5);
});
f.prototype.hush = function() {
  return v;
};
const $p = l(
  "palindrome",
  function(t) {
    return t.lastOf(2, gn);
  },
  !0,
  !0
), { juxBy: Rp, juxby: Lp } = l(["juxBy", "juxby"], function(t, e, n) {
  t /= 2;
  const s = function(c, u, a) {
    return u in c ? c[u] : a;
  }, r = n.withValue((c) => Object.assign({}, c, { pan: s(c, "pan", 0.5) - t })), o = e(n.withValue((c) => Object.assign({}, c, { pan: s(c, "pan", 0.5) + t })));
  return B(r, o).setSteps(j ? U(r._steps, o._steps) : void 0);
}), Ip = l("jux", function(t, e) {
  return e._juxBy(1, t, e);
}), { echoWith: Hp, echowith: Vp, stutWith: Fp, stutwith: Dp } = l(
  ["echoWith", "echowith", "stutWith", "stutwith"],
  function(t, e, n, s) {
    return B(...ht(0, t - 1).map((r) => n(s.late(h(e).mul(r)), r)));
  }
), Gp = l("echo", function(t, e, n, s) {
  return s._echoWith(t, e, (r, o) => r.gain(Math.pow(n, o)));
}), Qp = l("stut", function(t, e, n, s) {
  return s._echoWith(t, n, (r, o) => r.gain(Math.pow(e, o)));
}), bn = l("applyN", function(t, e, n) {
  let s = n;
  for (let r = 0; r < t; r++)
    s = e(s);
  return s;
}), Up = l(["plyWith", "plywith"], function(t, e, n) {
  const s = n.fmap((r) => at(...ht(0, t - 1).map((o) => bn(o, e, r)))._fast(t)).squeezeJoin();
  return j && (s._steps = h(t).mulmaybe(n._steps)), s;
}), Kp = l(["plyForEach", "plyforeach"], function(t, e, n) {
  const s = n.fmap((r) => at(at(z(r), ...ht(1, t - 1).map((o) => e(z(r), o))))._fast(t)).squeezeJoin();
  return j && (s._steps = h(t).mulmaybe(n._steps)), s;
}), Tt = function(t, e, n = !1) {
  return t = h(t), K(
    ...ht(0, t.sub(1)).map(
      (s) => n ? e.late(h(s).div(t)) : e.early(h(s).div(t))
    )
  );
}, Xp = l(
  "iter",
  function(t, e) {
    return Tt(t, e, !1);
  },
  !0,
  !0
), { iterBack: Yp, iterback: Zp } = l(
  ["iterBack", "iterback"],
  function(t, e) {
    return Tt(t, e, !0);
  },
  !0,
  !0
), { repeatCycles: tf } = l(
  "repeatCycles",
  function(t, e) {
    return new f(function(n) {
      const s = n.span.begin.sam(), r = s.div(t).sam(), o = s.sub(r);
      return n = n.withSpan((c) => c.withTime((u) => u.sub(o))), e.query(n).map((c) => c.withSpan((u) => u.withTime((a) => a.add(o))));
    }).splitQueries();
  },
  !0,
  !0
), Ct = function(t, e, n, s = !1, r = !1) {
  const o = Array(t - 1).fill(!1);
  o.unshift(!0);
  const c = Tt(t, G(...o), !s);
  return r || (n = n.repeatCycles(t)), n.when(c, e);
}, { chunk: ef, slowchunk: nf, slowChunk: sf } = l(
  ["chunk", "slowchunk", "slowChunk"],
  function(t, e, n) {
    return Ct(t, e, n, !1, !1);
  },
  !0,
  !0
), { chunkBack: rf, chunkback: of } = l(
  ["chunkBack", "chunkback"],
  function(t, e, n) {
    return Ct(t, e, n, !0);
  },
  !0,
  !0
), { fastchunk: cf, fastChunk: uf } = l(
  ["fastchunk", "fastChunk"],
  function(t, e, n) {
    return Ct(t, e, n, !1, !0);
  },
  !0,
  !0
), { chunkinto: af, chunkInto: lf } = l(["chunkinto", "chunkInto"], function(t, e, n) {
  return n.into(W(!0, ...Array(t - 1).fill(!1))._iterback(t), e);
}), { chunkbackinto: pf, chunkBackInto: ff } = l(["chunkbackinto", "chunkBackInto"], function(t, e, n) {
  return n.into(
    W(!0, ...Array(t - 1).fill(!1))._iter(t)._early(1),
    e
  );
}), hf = l(
  "bypass",
  function(t, e) {
    return t = !!parseInt(t), t ? v : e;
  },
  !0,
  !0
), { ribbon: df, rib: mf } = l(
  ["ribbon", "rib"],
  (t, e, n) => n.early(t).restart(z(1).slow(e))
), yf = l("hsla", (t, e, n, s, r) => r.color(`hsla(${t}turn,${e * 100}%,${n * 100}%,${s})`)), wf = l("hsl", (t, e, n, s) => s.color(`hsl(${t}turn,${e * 100}%,${n * 100}%)`));
f.prototype.tag = function(t) {
  return this.withContext((e) => ({ ...e, tags: (e.tags || []).concat([t]) }));
};
const gf = l("filter", (t, e) => e.withHaps((n) => n.filter(t))), bf = l("filterWhen", (t, e) => e.filter((n) => t(n.whole.begin))), _f = l(
  "within",
  (t, e, n, s) => B(
    n(s.filterWhen((r) => r.cyclePos() >= t && r.cyclePos() <= e)),
    s.filterWhen((r) => r.cyclePos() < t || r.cyclePos() > e)
  )
);
f.prototype.stepJoin = function() {
  const t = this, e = N(...Lt(It(t.queryArc(0, 1))))._steps, n = function(s) {
    const o = t.early(s.span.begin.sam()).query(s.setSpan(new x(h(0), h(1))));
    return N(...Lt(It(o))).query(s);
  };
  return new f(n, e);
};
f.prototype.stepBind = function(t) {
  return this.fmap(t).stepJoin();
};
function Lt(t) {
  const e = t.filter((o, c) => c.hasSteps).reduce((o, c) => o.add(c), h(0)), n = it(t.map((o, c) => c._steps)).reduce(
    (o, c) => o.add(c),
    h(0)
  ), s = e.eq(0) ? void 0 : n.div(e);
  function r(o, c) {
    return c._steps === void 0 ? [o.mulmaybe(s), c] : [c._steps, c];
  }
  return t.map((o) => r(...o));
}
function It(t) {
  const e = D(t.map((r) => [r.part.begin, r.part.end])), n = Ke([h(0), h(1), ...e]);
  return Ie(n).map((r) => [
    r[1].sub(r[0]),
    B(..._n(new x(...r), t).map((o) => o.value.withHap((c) => c.setContext(c.combineContext(o)))))
  ]);
}
function _n(t, e) {
  return it(e.map((n) => vn(t, n)));
}
function vn(t, e) {
  const n = t.intersection(e.part);
  if (n != null)
    return new C(e.whole, n, e.value, e.context);
}
const kn = l("pace", function(t, e) {
  return e._steps === void 0 ? e : e._steps.eq(h(0)) ? R : e._fast(h(t).div(e._steps)).setSteps(t);
});
function qn(t, ...e) {
  const n = e.map((r) => vt(r));
  if (n.length == 0)
    return v;
  t == 0 && (t = n[0][1]);
  const s = [];
  for (const r of n)
    r[1] != 0 && (t == r[1] ? s.push(r[0]) : s.push(r[0]._fast(h(t).div(h(r[1])))));
  return B(...s);
}
function xt(...t) {
  if (Array.isArray(t[0]))
    return qn(0, ...t);
  if (t = t.filter((s) => s.hasSteps), t.length == 0)
    return v;
  const e = U(...t.map((s) => s._steps));
  if (e.eq(h(0)))
    return R;
  const n = B(...t.map((s) => s.pace(e)));
  return n._steps = e, n;
}
function N(...t) {
  if (t.length === 0)
    return R;
  const e = (c) => Array.isArray(c) ? c : [c._steps ?? 1, c];
  if (t = t.map(e), t.find((c) => c[0] === void 0)) {
    const c = t.map((a) => a[0]).filter((a) => a !== void 0);
    if (c.length === 0)
      return W(...t.map((a) => a[1]));
    if (c.length === t.length)
      return R;
    const u = c.reduce((a, p) => a.add(p), h(0)).div(c.length);
    for (let a of t)
      a[0] === void 0 && (a[0] = u);
  }
  if (t.length == 1)
    return d(t[0][1]).withSteps((u) => t[0][0]);
  const n = t.map((c) => c[0]).reduce((c, u) => c.add(u), h(0));
  let s = h(0);
  const r = [];
  for (const [c, u] of t) {
    if (h(c).eq(0))
      continue;
    const a = s.add(c);
    r.push(d(u)._compress(s.div(n), a.div(n))), s = a;
  }
  const o = B(...r);
  return o._steps = n, o;
}
function An(...t) {
  t = t.map((r) => Array.isArray(r) ? r.map(d) : [d(r)]);
  const e = U(...t.map((r) => h(r.length)));
  let n = [];
  for (let r = 0; r < e; ++r)
    n.push(...t.map((o) => o.length == 0 ? v : o[r % o.length]));
  n = n.filter((r) => r.hasSteps && r._steps > 0);
  const s = n.reduce((r, o) => r.add(o._steps), h(0));
  return n = N(...n), n._steps = s, n;
}
const Sn = X("take", function(t, e) {
  if (!e.hasSteps || e._steps.lte(0) || (t = h(t), t.eq(0)))
    return R;
  const n = t < 0;
  n && (t = t.abs());
  const s = t.div(e._steps);
  return s.lte(0) ? R : s.gte(1) ? e : n ? e.zoom(h(1).sub(s), 1) : e.zoom(0, s);
}), Tn = X("drop", function(t, e) {
  return e.hasSteps ? (t = h(t), t.lt(0) ? e.take(e._steps.add(t)) : e.take(h(0).sub(e._steps.sub(t)))) : R;
}), Cn = X("extend", function(t, e) {
  return e.fast(t).expand(t);
}), vf = X("replicate", function(t, e) {
  return e.repeatCycles(t).fast(t).expand(t);
}), xn = X("expand", function(t, e) {
  return e.withSteps((n) => n.mul(h(t)));
}), Bn = X("contract", function(t, e) {
  return e.withSteps((n) => n.div(h(t)));
});
f.prototype.shrinklist = function(t) {
  const e = this;
  if (!e.hasSteps)
    return [e];
  let [n, s] = Array.isArray(t) ? t : [t, e._steps];
  if (n = h(n), s === 0 || n === 0)
    return [e];
  const r = n > 0, o = [];
  if (r) {
    const c = h(1).div(e._steps).mul(n);
    for (let u = 0; u < s; ++u) {
      const a = c.mul(u);
      if (a.gt(1))
        break;
      o.push([a, 1]);
    }
  } else {
    n = h(0).sub(n);
    const c = h(1).div(e._steps).mul(n);
    for (let u = 0; u < s; ++u) {
      const a = h(1).sub(c.mul(u));
      if (a.lt(0))
        break;
      o.push([h(0), a]);
    }
  }
  return o.map((c) => e.zoom(...c));
};
const On = (t, e) => e.shrinklist(t), zn = l(
  "shrink",
  function(t, e) {
    if (!e.hasSteps)
      return R;
    const n = e.shrinklist(t), s = N(...n);
    return s._steps = n.reduce((r, o) => r.add(o._steps), h(0)), s;
  },
  !0,
  !1,
  (t) => t.stepJoin()
), kf = l(
  "grow",
  function(t, e) {
    if (!e.hasSteps)
      return R;
    const n = e.shrinklist(h(0).sub(t));
    n.reverse();
    const s = N(...n);
    return s._steps = n.reduce((r, o) => r.add(o._steps), h(0)), s;
  },
  !0,
  !1,
  (t) => t.stepJoin()
), En = function(t, ...e) {
  return t.tour(...e);
};
f.prototype.tour = function(...t) {
  return N(
    ...[].concat(
      ...t.map((e, n) => [...t.slice(0, t.length - n), this, ...t.slice(t.length - n)]),
      this,
      ...t
    )
  );
};
const Mn = function(...t) {
  t = t.filter((s) => s.hasSteps);
  const e = K(...t.map((s) => s._slow(s._steps))), n = U(...t.map((s) => s._steps));
  return e._fast(n).setSteps(n);
}, qf = N, Pn = N, Af = N, Sf = An, Tf = xt;
f.prototype.s_polymeter = f.prototype.polymeter;
const Cf = zn;
f.prototype.s_taper = f.prototype.shrink;
const xf = On;
f.prototype.s_taperlist = f.prototype.shrinklist;
const Bf = Sn;
f.prototype.s_add = f.prototype.take;
const Of = Tn;
f.prototype.s_sub = f.prototype.drop;
const zf = xn;
f.prototype.s_expand = f.prototype.expand;
const Ef = Cn;
f.prototype.s_extend = f.prototype.extend;
const Mf = Bn;
f.prototype.s_contract = f.prototype.contract;
const Pf = En;
f.prototype.s_tour = f.prototype.tour;
const Jf = Mn;
f.prototype.s_zip = f.prototype.zip;
const jf = kn;
f.prototype.steps = f.prototype.pace;
const Nf = l("chop", function(t, e) {
  const s = Array.from({ length: t }, (c, u) => u).map((c) => ({ begin: c / t, end: (c + 1) / t })), r = function(c, u) {
    if ("begin" in c && "end" in c && c.begin !== void 0 && c.end !== void 0) {
      const a = c.end - c.begin;
      u = { begin: c.begin + u.begin * a, end: c.begin + u.end * a };
    }
    return Object.assign({}, c, u);
  }, o = function(c) {
    return G(s.map((u) => r(c, u)));
  };
  return e.squeezeBind(o).setSteps(j ? h(t).mulmaybe(e._steps) : void 0);
}), Wf = l("striate", function(t, e) {
  const s = Array.from({ length: t }, (o, c) => c).map((o) => ({ begin: o / t, end: (o + 1) / t })), r = K(...s);
  return e.set(r)._fast(t).setSteps(j ? h(t).mulmaybe(e._steps) : void 0);
}), te = function(t, e, n = 0.5) {
  return e.speed(1 / t * n).unit("c").slow(t);
}, Jn = l(
  "slice",
  function(t, e, n) {
    return t.innerBind(
      (s) => e.outerBind(
        (r) => n.outerBind((o) => {
          o = o instanceof Object ? o : { s: o };
          const c = Array.isArray(s) ? s[r] : r / s, u = Array.isArray(s) ? s[r + 1] : (r + 1) / s;
          return z({ begin: c, end: u, _slices: s, ...o });
        })
      )
    ).setSteps(e._steps);
  },
  !1
  // turns off auto-patternification
);
f.prototype.onTriggerTime = function(t) {
  return this.onTrigger((e, n, s, r) => {
    const o = r - n;
    window.setTimeout(() => {
      t(e);
    }, o * 1e3);
  }, !1);
};
const $f = l(
  "splice",
  function(t, e, n) {
    const s = Jn(t, e, n);
    return new f((r) => {
      const o = r.controls._cps || 1;
      return s.query(r).map(
        (u) => u.withValue((a) => ({
          speed: o / a._slices / u.whole.duration * (a.speed || 1),
          unit: "c",
          ...a
        }))
      );
    }).setSteps(e._steps);
  },
  !1
  // turns off auto-patternification
), { loopAt: Rf, loopat: Lf } = l(["loopAt", "loopat"], function(t, e) {
  const n = e._steps ? e._steps.div(t) : void 0;
  return new f((s) => te(t, e, s.controls._cps).query(s), n);
}), If = l(
  "fit",
  (t) => t.withHaps(
    (e, n) => e.map(
      (s) => s.withValue((r) => {
        const o = ("end" in r ? r.end : 1) - ("begin" in r ? r.begin : 0);
        return {
          ...r,
          speed: (n.controls._cps || 1) / s.whole.duration * o,
          unit: "c"
        };
      })
    )
  )
), { loopAtCps: Hf, loopatcps: Vf } = l(["loopAtCps", "loopatcps"], function(t, e, n) {
  return te(t, n, e);
}), Ff = (t) => z(1).withValue(() => d(t())).innerJoin();
let Ht = (t) => t < 0.5 ? 1 : 1 - (t - 0.5) / 0.5, jn = (t, e, n) => {
  e = d(e), t = d(t), n = d(n);
  let s = e.fmap((o) => ({ gain: Ht(o) })), r = e.fmap((o) => ({ gain: Ht(1 - o) }));
  return B(t.mul(s), n.mul(r));
};
f.prototype.xfade = function(t, e) {
  return jn(this, t, e);
};
const Nn = (t) => (e, n, s) => {
  e = h(e).mod(n), n = h(n);
  const r = e.div(n), o = e.add(1).div(n);
  return t(s.fmap((c) => z(c)._compress(r, o)));
}, { beat: Df } = l(
  ["beat"],
  Nn((t) => t.innerJoin())
), ee = (t, e, n) => {
  n = h(n);
  const s = h(1).div(t.length), r = (u) => {
    const a = [];
    for (const [p, m] of u.entries())
      m && a.push([h(p).div(u.length), m]);
    return a;
  }, o = At(
    ([u, a], [p, m]) => {
      const y = n.mul(p - u).add(u), _ = y.add(s);
      return new x(y, _);
    },
    r(t),
    r(e)
  );
  function c(u) {
    const a = u.span.begin.sam(), p = u.span.cycleArc(), m = [];
    for (const y of o) {
      const _ = y.intersection(p);
      _ !== void 0 && m.push(
        new C(
          y.withTime((g) => g.add(a)),
          _.withTime((g) => g.add(a)),
          !0
        )
      );
    }
    return m;
  }
  return new f(c).splitQueries();
}, Gf = (t, e, n) => (t = d(t), e = d(e), n = d(n), t.innerBind((s) => e.innerBind((r) => n.innerBind((o) => ee(s, r, o))))), Wn = ["scurve", "soft", "hard", "cubic", "diode", "asym", "fold", "sinefold", "chebyshev"];
for (const t of Wn)
  f.prototype[t] = function(e) {
    const n = d(e).fmap((s) => Array.isArray(s) ? [...s, t] : [s, 1, t]);
    return this.distort(n);
  };
function Bt(t) {
  let e = Array.isArray(t);
  t = e ? t : [t];
  const n = t[0], s = (o) => {
    let c;
    if (typeof o == "object" && o.value !== void 0 && (c = { ...o }, o = o.value, delete c.value), e && Array.isArray(o)) {
      const u = c || {};
      return o.forEach((a, p) => {
        p < t.length && (u[t[p]] = a);
      }), u;
    } else return c ? (c[n] = o, c) : { [n]: o };
  }, r = function(o, c) {
    return c ? typeof o > "u" ? c.fmap(s) : c.set(d(o).withValue(s)) : d(o).withValue(s);
  };
  return f.prototype[n] = function(o) {
    return r(o, this);
  }, r;
}
const rt = /* @__PURE__ */ new Map();
function $n(t) {
  return rt.has(t);
}
function i(t, ...e) {
  const n = Array.isArray(t) ? t[0] : t;
  let s = {};
  return s[n] = Bt(t), rt.set(n, n), e.forEach((r) => {
    s[r] = s[n], rt.set(r, n), f.prototype[r] = f.prototype[n];
  }), s;
}
const { s: Rn, sound: Ln } = i(["s", "n", "gain"], "sound"), { wt: In, wavetablePosition: Hn } = i("wt", "wavetablePosition"), { wtenv: Vn } = i("wtenv"), { wtattack: Fn, wtatt: Dn } = i("wtattack", "wtatt"), { wtdecay: Gn, wtdec: Qn } = i("wtdecay", "wtdec"), { wtsustain: Un, wtsus: Kn } = i("wtsustain", "wtsus"), { wtrelease: Xn, wtrel: Yn } = i("wtrelease", "wtrel"), { wtrate: Zn } = i("wtrate"), { wtsync: ts } = i("wtsync"), { wtdepth: es } = i("wtdepth"), { wtshape: ns } = i("wtshape"), { wtdc: ss } = i("wtdc"), { wtskew: rs } = i("wtskew"), { warp: os, wavetableWarp: is } = i("warp", "wavetableWarp"), { warpattack: cs, warpatt: us } = i("warpattack", "warpatt"), { warpdecay: as, warpdec: ls } = i("warpdecay", "warpdec"), { warpsustain: ps, warpsus: fs } = i("warpsustain", "warpsus"), { warprelease: hs, warprel: ds } = i("warprelease", "warprel"), { warprate: ms } = i("warprate"), { warpdepth: ys } = i("warpdepth"), { warpshape: ws } = i("warpshape"), { warpdc: gs } = i("warpdc"), { warpskew: bs } = i("warpskew"), { warpmode: _s, wavetableWarpMode: vs } = i("warpmode", "wavetableWarpMode"), { wtphaserand: ks, wavetablePhaseRand: qs } = i("wtphaserand", "wavetablePhaseRand"), { warpenv: As } = i("warpenv"), { warpsync: Ss } = i("warpsync"), { source: Ts, src: Cs } = i("source", "src"), { n: xs } = i("n"), { note: Bs } = i(["note", "n"]), { accelerate: Os } = i("accelerate"), { velocity: zs } = i("velocity"), { gain: Es } = i("gain"), { postgain: Ms } = i("postgain"), { amp: Ps } = i("amp"), { attack: Js, att: js } = i("attack", "att"), { fmh: Ns } = i(["fmh", "fmi"], "fmh"), { fmi: Ws, fm: $s } = i(["fmi", "fmh"], "fm"), { fmenv: Rs } = i("fmenv"), { fmattack: Ls } = i("fmattack"), { fmwave: Is } = i("fmwave"), { fmdecay: Hs } = i("fmdecay"), { fmsustain: Vs } = i("fmsustain"), { fmrelease: Fs } = i("fmrelease"), { fmvelocity: Ds } = i("fmvelocity"), { bank: Gs } = i("bank"), { chorus: Qs } = i("chorus"), { analyze: Us } = i("analyze"), { fft: Ks } = i("fft"), { decay: Xs, dec: Ys } = i("decay", "dec"), { sustain: Zs, sus: tr } = i("sustain", "sus"), { release: er, rel: nr } = i("release", "rel"), { hold: sr } = i("hold"), { bandf: rr, bpf: or, bp: ir } = i(["bandf", "bandq", "bpenv"], "bpf", "bp"), { bandq: cr, bpq: ur } = i("bandq", "bpq"), { begin: ar } = i("begin"), { end: lr } = i("end"), { loop: pr } = i("loop"), { loopBegin: fr, loopb: hr } = i("loopBegin", "loopb"), { loopEnd: dr, loope: mr } = i("loopEnd", "loope"), { crush: yr } = i("crush"), { coarse: wr } = i("coarse"), { tremolo: gr } = i(["tremolo", "tremolodepth", "tremoloskew", "tremolophase"], "trem"), { tremolosync: br } = i(
  ["tremolosync", "tremolodepth", "tremoloskew", "tremolophase"],
  "tremsync"
), { tremolodepth: _r } = i("tremolodepth", "tremdepth"), { tremoloskew: vr } = i("tremoloskew", "tremskew"), { tremolophase: kr } = i("tremolophase", "tremphase"), { tremoloshape: qr } = i("tremoloshape", "tremshape"), { drive: Ar } = i("drive"), { duck: Sr } = i("duckorbit", "duck"), { duckdepth: Tr } = i("duckdepth"), { duckonset: Cr } = i("duckonset", "duckons"), { duckattack: xr } = i("duckattack", "duckatt"), { byteBeatExpression: Br, bbexpr: Or } = i("byteBeatExpression", "bbexpr"), { byteBeatStartTime: zr, bbst: Er } = i("byteBeatStartTime", "bbst"), { channels: Mr, ch: Pr } = i("channels", "ch"), { pw: Jr } = i(["pw", "pwrate", "pwsweep"]), { pwrate: jr } = i("pwrate"), { pwsweep: Nr } = i("pwsweep"), { phaserrate: Wr, ph: $r, phaser: Rr } = i(
  ["phaserrate", "phaserdepth", "phasercenter", "phasersweep"],
  "ph",
  "phaser"
), { phasersweep: Lr, phs: Ir } = i("phasersweep", "phs"), { phasercenter: Hr, phc: Vr } = i("phasercenter", "phc"), { phaserdepth: Fr, phd: Dr, phasdp: Gr } = i("phaserdepth", "phd", "phasdp"), { channel: Qr } = i("channel"), { cut: Ur } = i("cut"), { cutoff: Kr, ctf: Xr, lpf: Yr, lp: Zr } = i(["cutoff", "resonance", "lpenv"], "ctf", "lpf", "lp"), { lpenv: to, lpe: eo } = i("lpenv", "lpe"), { hpenv: no, hpe: so } = i("hpenv", "hpe"), { bpenv: ro, bpe: oo } = i("bpenv", "bpe"), { lpattack: io, lpa: co } = i("lpattack", "lpa"), { hpattack: uo, hpa: ao } = i("hpattack", "hpa"), { bpattack: lo, bpa: po } = i("bpattack", "bpa"), { lpdecay: fo, lpd: ho } = i("lpdecay", "lpd"), { hpdecay: mo, hpd: yo } = i("hpdecay", "hpd"), { bpdecay: wo, bpd: go } = i("bpdecay", "bpd"), { lpsustain: bo, lps: _o } = i("lpsustain", "lps"), { hpsustain: vo, hps: ko } = i("hpsustain", "hps"), { bpsustain: qo, bps: Ao } = i("bpsustain", "bps"), { lprelease: So, lpr: To } = i("lprelease", "lpr"), { hprelease: Co, hpr: xo } = i("hprelease", "hpr"), { bprelease: Bo, bpr: Oo } = i("bprelease", "bpr"), { ftype: zo } = i("ftype"), { fanchor: Eo } = i("fanchor"), { vib: Mo, vibrato: Po, v: Jo } = i(["vib", "vibmod"], "vibrato", "v"), { noise: jo } = i("noise"), { vibmod: No, vmod: Wo } = i(["vibmod", "vib"], "vmod"), { hcutoff: $o, hpf: Ro, hp: Lo } = i(["hcutoff", "hresonance", "hpenv"], "hpf", "hp"), { hresonance: Io, hpq: Ho } = i("hresonance", "hpq"), { resonance: Vo, lpq: Fo } = i("resonance", "lpq"), { djf: Do } = i("djf"), { delay: Go } = i(["delay", "delaytime", "delayfeedback"]), { delayfeedback: Qo, delayfb: Uo, dfb: Ko } = i("delayfeedback", "delayfb", "dfb"), { delayspeed: Xo } = i("delayspeed"), { delaytime: Yo, delayt: Zo, dt: ti } = i("delaytime", "delayt", "dt"), { delaysync: ei } = i("delaysync"), { lock: ni } = i("lock"), { detune: si, det: ri } = i("detune", "det"), { unison: oi } = i("unison"), { spread: ii } = i("spread"), { dry: ci } = i("dry"), { fadeTime: ui, fadeOutTime: ai } = i("fadeTime", "fadeOutTime"), { fadeInTime: li } = i("fadeInTime"), { freq: pi } = i("freq"), { pattack: fi, patt: hi } = i("pattack", "patt"), { pdecay: di, pdec: mi } = i("pdecay", "pdec"), { psustain: yi, psus: wi } = i("psustain", "psus"), { prelease: gi, prel: bi } = i("prelease", "prel"), { penv: _i } = i("penv"), { pcurve: vi } = i("pcurve"), { panchor: ki } = i("panchor"), { gate: qi, gat: Ai } = i("gate", "gat"), { leslie: Si } = i("leslie"), { lrate: Ti } = i("lrate"), { lsize: Ci } = i("lsize"), { activeLabel: xi } = i("activeLabel"), { label: Bi } = i(["label", "activeLabel"]), { degree: Oi } = i("degree"), { mtranspose: zi } = i("mtranspose"), { ctranspose: Ei } = i("ctranspose"), { harmonic: Mi } = i("harmonic"), { stepsPerOctave: Pi } = i("stepsPerOctave"), { octaveR: Ji } = i("octaveR"), { nudge: ji } = i("nudge"), { octave: Ni } = i("octave"), { orbit: Wi } = i("orbit"), { overgain: $i } = i("overgain"), { overshape: Ri } = i("overshape"), { pan: Li } = i("pan"), { panspan: Ii } = i("panspan"), { pansplay: Hi } = i("pansplay"), { panwidth: Vi } = i("panwidth"), { panorient: Fi } = i("panorient"), { rate: Di } = i("rate"), { slide: Gi } = i("slide"), { semitone: Qi } = i("semitone"), { voice: Ui } = i("voice"), { chord: Ki } = i("chord"), { dictionary: Xi, dict: Yi } = i("dictionary", "dict"), { anchor: Zi } = i("anchor"), { offset: tc } = i("offset"), { octaves: ec } = i("octaves"), { mode: nc } = i(["mode", "anchor"]), { room: sc } = i(["room", "size"]), { roomlp: rc, rlp: oc } = i("roomlp", "rlp"), { roomdim: ic, rdim: cc } = i("roomdim", "rdim"), { roomfade: uc, rfade: ac } = i("roomfade", "rfade"), { ir: lc, iresponse: pc } = i(["ir", "i"], "iresponse"), { irspeed: fc } = i("irspeed"), { irbegin: hc } = i("irbegin"), { roomsize: dc, size: mc, sz: yc, rsize: wc } = i("roomsize", "size", "sz", "rsize"), { shape: gc } = i(["shape", "shapevol"]), { distort: bc, dist: _c } = i(["distort", "distortvol", "distorttype"], "dist"), { distortvol: vc } = i("distortvol", "distvol"), { distorttype: kc } = i("distorttype", "disttype"), { compressor: qc } = i([
  "compressor",
  "compressorRatio",
  "compressorKnee",
  "compressorAttack",
  "compressorRelease"
]), { compressorKnee: Ac } = i("compressorKnee"), { compressorRatio: Sc } = i("compressorRatio"), { compressorAttack: Tc } = i("compressorAttack"), { compressorRelease: Cc } = i("compressorRelease"), { speed: ne } = i("speed"), { stretch: xc } = i("stretch"), { unit: Bc } = i("unit"), { squiz: Oc } = i("squiz"), { vowel: zc } = i("vowel"), { waveloss: Ec } = i("waveloss"), { density: Mc } = i("density"), { expression: Pc } = i("expression"), { sustainpedal: Jc } = i("sustainpedal"), { fshift: jc } = i("fshift"), { fshiftnote: Nc } = i("fshiftnote"), { fshiftphase: Wc } = i("fshiftphase"), { triode: $c } = i("triode"), { krush: Rc } = i("krush"), { kcutoff: Lc } = i("kcutoff"), { octer: Ic } = i("octer"), { octersub: Hc } = i("octersub"), { octersubsub: Vc } = i("octersubsub"), { ring: Fc } = i("ring"), { ringf: Dc } = i("ringf"), { ringdf: Gc } = i("ringdf"), { freeze: Qc } = i("freeze"), { xsdelay: Uc } = i("xsdelay"), { tsdelay: Kc } = i("tsdelay"), { real: Xc } = i("real"), { imag: Yc } = i("imag"), { enhance: Zc } = i("enhance"), { partials: tu } = i("partials"), { comb: eu } = i("comb"), { smear: nu } = i("smear"), { scram: su } = i("scram"), { binshift: ru } = i("binshift"), { hbrick: ou } = i("hbrick"), { lbrick: iu } = i("lbrick"), { frameRate: cu } = i("frameRate"), { frames: uu } = i("frames"), { hours: au } = i("hours"), { minutes: lu } = i("minutes"), { seconds: pu } = i("seconds"), { songPtr: fu } = i("songPtr"), { uid: hu } = i("uid"), { val: du } = i("val"), { cps: mu } = i("cps"), { clip: yu, legato: wu } = i("clip", "legato"), { duration: gu, dur: bu } = i("duration", "dur"), { zrand: _u } = i("zrand"), { curve: vu } = i("curve"), { deltaSlide: ku } = i("deltaSlide"), { pitchJump: qu } = i("pitchJump"), { pitchJumpTime: Au } = i("pitchJumpTime"), { lfo: Su, repeatTime: Tu } = i("lfo", "repeatTime"), { znoise: Cu } = i("znoise"), { zmod: xu } = i("zmod"), { zcrush: Bu } = i("zcrush"), { zdelay: Ou } = i("zdelay"), { zzfx: zu } = i("zzfx"), { color: Eu, colour: Mu } = i(["color", "colour"]);
let Pu = (...t) => t.reduce((e, n) => Object.assign(e, { [n]: Bt(n) }), {});
const Ju = l("adsr", (t, e) => {
  t = Array.isArray(t) ? t : [t];
  const [n, s, r, o] = t;
  return e.set({ attack: n, decay: s, sustain: r, release: o });
}), ju = l("ad", (t, e) => {
  t = Array.isArray(t) ? t : [t];
  const [n, s = n] = t;
  return e.attack(n).decay(s);
}), Nu = l("ds", (t, e) => {
  t = Array.isArray(t) ? t : [t];
  const [n, s = 0] = t;
  return e.set({ decay: n, sustain: s });
}), Wu = l("ar", (t, e) => {
  t = Array.isArray(t) ? t : [t];
  const [n, s = n] = t;
  return e.set({ attack: n, release: s });
}), { midichan: $u } = i("midichan"), { midimap: Ru } = i("midimap"), { midiport: Lu } = i("midiport"), { midicmd: Iu } = i("midicmd"), Hu = l("control", (t, e) => {
  if (!Array.isArray(t))
    throw new Error("control expects an array of [ccn, ccv]");
  const [n, s] = t;
  return e.ccn(n).ccv(s);
}), { ccn: Vu } = i("ccn"), { ccv: Fu } = i("ccv"), { ctlNum: Du } = i("ctlNum"), { nrpnn: Gu } = i("nrpnn"), { nrpv: Qu } = i("nrpv"), { progNum: Uu } = i("progNum"), Ku = l("sysex", (t, e) => {
  if (!Array.isArray(t))
    throw new Error("sysex expects an array of [id, data]");
  const [n, s] = t;
  return e.sysexid(n).sysexdata(s);
}), { sysexid: Xu } = i("sysexid"), { sysexdata: Yu } = i("sysexdata"), { midibend: Zu } = i("midibend"), { miditouch: ta } = i("miditouch"), { polyTouch: ea } = i("polyTouch"), { oschost: na } = i("oschost"), { oscport: sa } = i("oscport"), se = (t) => rt.has(t) ? rt.get(t) : t, ra = l("as", (t, e) => (t = Array.isArray(t) ? t : [t], e.fmap((n) => (n = Array.isArray(n) ? n : [n], n = Object.fromEntries(t.map((s, r) => [se(s), n[r]])), n)))), oa = l(
  "scrub",
  (t, e) => t.outerBind((n) => {
    Array.isArray(n) || (n = [n]);
    const [s, r = 1] = n;
    return e.begin(s).mul(ne(r)).clip(1);
  }),
  !1
), Qf = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  accelerate: Os,
  activeLabel: xi,
  ad: ju,
  adsr: Ju,
  amp: Ps,
  analyze: Us,
  anchor: Zi,
  ar: Wu,
  as: ra,
  att: js,
  attack: Js,
  bandf: rr,
  bandq: cr,
  bank: Gs,
  bbexpr: Or,
  bbst: Er,
  begin: ar,
  binshift: ru,
  bp: ir,
  bpa: po,
  bpattack: lo,
  bpd: go,
  bpdecay: wo,
  bpe: oo,
  bpenv: ro,
  bpf: or,
  bpq: ur,
  bpr: Oo,
  bprelease: Bo,
  bps: Ao,
  bpsustain: qo,
  byteBeatExpression: Br,
  byteBeatStartTime: zr,
  ccn: Vu,
  ccv: Fu,
  ch: Pr,
  channel: Qr,
  channels: Mr,
  chord: Ki,
  chorus: Qs,
  clip: yu,
  coarse: wr,
  color: Eu,
  colour: Mu,
  comb: eu,
  compressor: qc,
  compressorAttack: Tc,
  compressorKnee: Ac,
  compressorRatio: Sc,
  compressorRelease: Cc,
  control: Hu,
  cps: mu,
  createParam: Bt,
  createParams: Pu,
  crush: yr,
  ctf: Xr,
  ctlNum: Du,
  ctranspose: Ei,
  curve: vu,
  cut: Ur,
  cutoff: Kr,
  dec: Ys,
  decay: Xs,
  degree: Oi,
  delay: Go,
  delayfb: Uo,
  delayfeedback: Qo,
  delayspeed: Xo,
  delaysync: ei,
  delayt: Zo,
  delaytime: Yo,
  deltaSlide: ku,
  density: Mc,
  det: ri,
  detune: si,
  dfb: Ko,
  dict: Yi,
  dictionary: Xi,
  dist: _c,
  distort: bc,
  distorttype: kc,
  distortvol: vc,
  djf: Do,
  drive: Ar,
  dry: ci,
  ds: Nu,
  dt: ti,
  duck: Sr,
  duckattack: xr,
  duckdepth: Tr,
  duckonset: Cr,
  dur: bu,
  duration: gu,
  end: lr,
  enhance: Zc,
  expression: Pc,
  fadeInTime: li,
  fadeOutTime: ai,
  fadeTime: ui,
  fanchor: Eo,
  fft: Ks,
  fm: $s,
  fmattack: Ls,
  fmdecay: Hs,
  fmenv: Rs,
  fmh: Ns,
  fmi: Ws,
  fmrelease: Fs,
  fmsustain: Vs,
  fmvelocity: Ds,
  fmwave: Is,
  frameRate: cu,
  frames: uu,
  freeze: Qc,
  freq: pi,
  fshift: jc,
  fshiftnote: Nc,
  fshiftphase: Wc,
  ftype: zo,
  gain: Es,
  gat: Ai,
  gate: qi,
  getControlName: se,
  harmonic: Mi,
  hbrick: ou,
  hcutoff: $o,
  hold: sr,
  hours: au,
  hp: Lo,
  hpa: ao,
  hpattack: uo,
  hpd: yo,
  hpdecay: mo,
  hpe: so,
  hpenv: no,
  hpf: Ro,
  hpq: Ho,
  hpr: xo,
  hprelease: Co,
  hps: ko,
  hpsustain: vo,
  hresonance: Io,
  imag: Yc,
  ir: lc,
  irbegin: hc,
  iresponse: pc,
  irspeed: fc,
  isControlName: $n,
  kcutoff: Lc,
  krush: Rc,
  label: Bi,
  lbrick: iu,
  legato: wu,
  leslie: Si,
  lfo: Su,
  lock: ni,
  loop: pr,
  loopBegin: fr,
  loopEnd: dr,
  loopb: hr,
  loope: mr,
  lp: Zr,
  lpa: co,
  lpattack: io,
  lpd: ho,
  lpdecay: fo,
  lpe: eo,
  lpenv: to,
  lpf: Yr,
  lpq: Fo,
  lpr: To,
  lprelease: So,
  lps: _o,
  lpsustain: bo,
  lrate: Ti,
  lsize: Ci,
  midibend: Zu,
  midichan: $u,
  midicmd: Iu,
  midimap: Ru,
  midiport: Lu,
  miditouch: ta,
  minutes: lu,
  mode: nc,
  mtranspose: zi,
  n: xs,
  noise: jo,
  note: Bs,
  nrpnn: Gu,
  nrpv: Qu,
  nudge: ji,
  octave: Ni,
  octaveR: Ji,
  octaves: ec,
  octer: Ic,
  octersub: Hc,
  octersubsub: Vc,
  offset: tc,
  orbit: Wi,
  oschost: na,
  oscport: sa,
  overgain: $i,
  overshape: Ri,
  pan: Li,
  panchor: ki,
  panorient: Fi,
  panspan: Ii,
  pansplay: Hi,
  panwidth: Vi,
  partials: tu,
  patt: hi,
  pattack: fi,
  pcurve: vi,
  pdec: mi,
  pdecay: di,
  penv: _i,
  ph: $r,
  phasdp: Gr,
  phaser: Rr,
  phasercenter: Hr,
  phaserdepth: Fr,
  phaserrate: Wr,
  phasersweep: Lr,
  phc: Vr,
  phd: Dr,
  phs: Ir,
  pitchJump: qu,
  pitchJumpTime: Au,
  polyTouch: ea,
  postgain: Ms,
  prel: bi,
  prelease: gi,
  progNum: Uu,
  psus: wi,
  psustain: yi,
  pw: Jr,
  pwrate: jr,
  pwsweep: Nr,
  rate: Di,
  rdim: cc,
  real: Xc,
  registerControl: i,
  rel: nr,
  release: er,
  repeatTime: Tu,
  resonance: Vo,
  rfade: ac,
  ring: Fc,
  ringdf: Gc,
  ringf: Dc,
  rlp: oc,
  room: sc,
  roomdim: ic,
  roomfade: uc,
  roomlp: rc,
  roomsize: dc,
  rsize: wc,
  s: Rn,
  scram: su,
  scrub: oa,
  seconds: pu,
  semitone: Qi,
  shape: gc,
  size: mc,
  slide: Gi,
  smear: nu,
  songPtr: fu,
  sound: Ln,
  source: Ts,
  speed: ne,
  spread: ii,
  squiz: Oc,
  src: Cs,
  stepsPerOctave: Pi,
  stretch: xc,
  sus: tr,
  sustain: Zs,
  sustainpedal: Jc,
  sysex: Ku,
  sysexdata: Yu,
  sysexid: Xu,
  sz: yc,
  tremolo: gr,
  tremolodepth: _r,
  tremolophase: kr,
  tremoloshape: qr,
  tremoloskew: vr,
  tremolosync: br,
  triode: $c,
  tsdelay: Kc,
  uid: hu,
  unison: oi,
  unit: Bc,
  v: Jo,
  val: du,
  velocity: zs,
  vib: Mo,
  vibmod: No,
  vibrato: Po,
  vmod: Wo,
  voice: Ui,
  vowel: zc,
  warp: os,
  warpatt: us,
  warpattack: cs,
  warpdc: gs,
  warpdec: ls,
  warpdecay: as,
  warpdepth: ys,
  warpenv: As,
  warpmode: _s,
  warprate: ms,
  warprel: ds,
  warprelease: hs,
  warpshape: ws,
  warpskew: bs,
  warpsus: fs,
  warpsustain: ps,
  warpsync: Ss,
  waveloss: Ec,
  wavetablePhaseRand: qs,
  wavetablePosition: Hn,
  wavetableWarp: is,
  wavetableWarpMode: vs,
  wt: In,
  wtatt: Dn,
  wtattack: Fn,
  wtdc: ss,
  wtdec: Qn,
  wtdecay: Gn,
  wtdepth: es,
  wtenv: Vn,
  wtphaserand: ks,
  wtrate: Zn,
  wtrel: Yn,
  wtrelease: Xn,
  wtshape: ns,
  wtskew: rs,
  wtsus: Kn,
  wtsustain: Un,
  wtsync: ts,
  xsdelay: Uc,
  zcrush: Bu,
  zdelay: Ou,
  zmod: xu,
  znoise: Cu,
  zrand: _u,
  zzfx: zu
}, Symbol.toStringTag, { value: "Module" })), ia = function(t, e) {
  const [n, s] = t, [r, o] = e, [c, u] = Kt(s, r);
  return [
    [s, n - s],
    [At((a, p) => a.concat(p), c, o), u]
  ];
}, ca = function(t, e) {
  const [n, s] = t, [r, o] = e, [c, u] = Kt(n, o);
  return [
    [n, s - n],
    [At((p, m) => p.concat(m), r, c), u]
  ];
}, re = function(t, e) {
  const [n, s] = t;
  return Math.min(n, s) <= 1 ? [t, e] : re(...n > s ? ia(t, e) : ca(t, e));
}, oe = function(t, e) {
  const n = t < 0, s = Math.abs(t), r = e - s, o = Array(s).fill([1]), c = Array(r).fill([0]), u = re([s, r], [o, c]), a = D(u[1][0]).concat(D(u[1][1]));
  return n ? a.map((p) => 1 - p) : a;
}, dt = function(t, e, n) {
  const s = oe(t, e);
  return n ? $e(s, -n) : s;
}, Uf = l("euclid", function(t, e, n) {
  return n.struct(dt(t, e, 0));
}), Kf = l("e", function(t, e) {
  Array.isArray(t) || (t = [t]);
  const [n, s = n, r = 0] = t;
  return e.struct(dt(n, s, r));
}), { euclidrot: Xf, euclidRot: Yf } = l(["euclidrot", "euclidRot"], function(t, e, n, s) {
  return s.struct(dt(t, e, n));
}), ie = function(t, e, n, s) {
  if (t < 1)
    return v;
  const o = dt(t, e, 0).join("").split("1").slice(1).map((c) => [c.length + 1, !0]);
  return s.struct(Pn(...o)).late(h(n).div(e));
}, Zf = l(["euclidLegato"], function(t, e, n) {
  return ie(t, e, 0, n);
}), th = l(["euclidLegatoRot"], function(t, e, n, s) {
  return ie(t, e, n, s);
}), { euclidish: eh, eish: nh } = l(["euclidish", "eish"], function(t, e, n, s) {
  const r = ee(oe(t, e), new Array(t).fill(1), n);
  return s.struct(r).setSteps(e);
});
function ua(t, e, n = 0.05, s = 0.1, r = 0.1, o = globalThis.setInterval, c = globalThis.clearInterval, u = !0) {
  let a = 0, p = 0, m = 10 ** 4, y = 0.01;
  const _ = (J) => n = J(n);
  r = r || s / 2;
  const g = () => {
    const J = t(), Q = J + s + r;
    for (p === 0 && (p = J + y); p < Q; )
      p = u ? Math.round(p * m) / m : p, e(p, n, a, J), p += n, a++;
  };
  let k;
  const q = () => {
    M(), g(), k = o(g, s * 1e3);
  }, M = () => {
    k !== void 0 && c(k), k = void 0;
  };
  return { setDuration: _, start: q, stop: () => {
    a = 0, p = 0, M();
  }, pause: () => M(), duration: n, interval: s, getPhase: () => p, minLatency: y };
}
function sh(t) {
  return new f((e) => [new C(void 0, e.span, t)]);
}
const I = (t) => {
  const e = (n) => [new C(void 0, n.span, t(n.span.begin))];
  return new f(e);
}, mt = I((t) => t % 1), ce = mt.toBipolar(), Ot = I((t) => 1 - t % 1), ue = Ot.toBipolar(), ae = I((t) => Math.sin(Math.PI * 2 * t)), aa = ae.fromBipolar(), rh = aa._early(h(1).div(4)), oh = ae._early(h(1).div(4)), la = I((t) => Math.floor(t * 2 % 2)), ih = la.toBipolar(), ch = W(mt, Ot), uh = W(ce, ue), ah = W(Ot, mt), lh = W(ue, ce), le = I(et);
let zt = 0, Et = 0;
typeof window < "u" && document.addEventListener("mousemove", (t) => {
  zt = t.clientY / document.body.clientHeight, Et = t.clientX / document.body.clientWidth;
});
const ph = I(() => zt), fh = I(() => zt), hh = I(() => Et), dh = I(() => Et), pe = (t) => {
  const e = t << 13 ^ t, n = e >> 17 ^ e;
  return n << 5 ^ n;
}, pa = (t) => t - Math.trunc(t), fe = (t) => pe(Math.trunc(pa(t / 300) * 536870912)), he = (t) => t % 536870912 / 536870912, ot = (t) => Math.abs(he(fe(t))), fa = (t, e) => {
  const n = [];
  for (let s = 0; s < e; ++s)
    n.push(he(t)), t = pe(t);
  return n;
}, ha = (t, e) => fa(fe(t), e), da = (t) => mt.range(0, t).round().segment(t), mh = (t) => {
  const e = d(t).log2(0).floor().add(1);
  return ma(t, e);
}, ma = (t, e = 16) => {
  e = d(e);
  const n = da(e).mul(-1).add(e.sub(1));
  return d(t).segment(e).brshift(n).band(z(1));
}, ya = (t) => I((e) => {
  const s = ha(e.floor().add(0.5), t).map((o, c) => [o, c]).sort((o, c) => (o[0] > c[0]) - (o[0] < c[0])).map((o) => o[1]), r = e.cyclePos().mul(t).floor() % t;
  return s[r];
})._segment(t), de = (t, e, n) => {
  const s = [...Array(e).keys()].map((r) => n.zoom(h(r).div(e), h(r + 1).div(e)));
  return t.fmap((r) => s[r].repeatCycles(e)._fast(e)).innerJoin();
}, yh = l("shuffle", (t, e) => de(ya(t), t, e)), wh = l("scramble", (t, e) => de(ye(t)._segment(t), t, e)), L = I(ot), gh = L.toBipolar(), me = (t) => L.fmap((e) => e < t), bh = (t) => d(t).fmap(me).innerJoin(), _h = me(0.5), ye = (t) => L.fmap((e) => Math.trunc(e * t)), vh = (t) => d(t).fmap(ye).innerJoin(), we = (t, e) => (e = e.map(d), e.length == 0 ? v : t.range(0, e.length).fmap((n) => {
  const s = Math.min(Math.max(Math.floor(n), 0), e.length - 1);
  return e[s];
})), Mt = (t, e) => we(t, e).outerJoin(), ge = (t, e) => we(t, e).innerJoin(), wa = (...t) => Mt(L, t), kh = (...t) => ge(L, t), qh = wa;
f.prototype.choose = function(...t) {
  return Mt(this, t);
};
f.prototype.choose2 = function(...t) {
  return Mt(this.fromBipolar(), t);
};
const ga = (...t) => ge(L.segment(1), t), Ah = ga, be = function(t, ...e) {
  const n = e.map((u) => d(u[0])), s = [];
  let r = z(0);
  for (const u of e)
    r = r.add(u[1]), s.push(r);
  const o = fn(s), c = function(u) {
    const a = r.mul(u);
    return o.fmap((p) => (m) => n[p.findIndex((y) => y > m, p)]).appLeft(a);
  };
  return t.bind(c);
}, ba = (...t) => be(...t).outerJoin(), Sh = (...t) => ba(L, ...t), _a = (...t) => be(L.segment(1), ...t).innerJoin(), Th = _a;
function va(t) {
  let e = Math.floor(t), n = e + 1;
  const s = (c) => 6 * c ** 5 - 15 * c ** 4 + 10 * c ** 3;
  return ((c) => (u) => (a) => u + s(c) * (a - u))(t - e)(ot(e))(ot(n));
}
const ka = (t) => t.fmap(va);
function qa(t) {
  const e = Math.floor(t), n = e + 1, s = ot(e), r = ot(n) + s, o = (t - e) / (n - e);
  return ((u, a, p) => u + (a - u) * p)(s, r, o) / 2;
}
const Aa = (t) => t.fmap(qa), Ch = ka(le.fmap((t) => Number(t))), xh = Aa(le.fmap((t) => Number(t))), Bh = l(
  "degradeByWith",
  (t, e, n) => n.fmap((s) => (r) => s).appLeft(t.filterValues((s) => s > e)),
  !0,
  !0
), Oh = l(
  "degradeBy",
  function(t, e) {
    return e._degradeByWith(L, t);
  },
  !0,
  !0
), zh = l("degrade", (t) => t._degradeBy(0.5), !0, !0), Eh = l(
  "undegradeBy",
  function(t, e) {
    return e._degradeByWith(
      L.fmap((n) => 1 - n),
      t
    );
  },
  !0,
  !0
), Mh = l("undegrade", (t) => t._undegradeBy(0.5), !0, !0), Ph = l("sometimesBy", function(t, e, n) {
  return d(t).fmap((s) => B(n._degradeBy(s), e(n._undegradeBy(1 - s)))).innerJoin();
}), Jh = l("sometimes", function(t, e) {
  return e._sometimesBy(0.5, t);
}), jh = l("someCyclesBy", function(t, e, n) {
  return d(t).fmap(
    (s) => B(
      n._degradeByWith(L._segment(1), s),
      e(n._degradeByWith(L.fmap((r) => 1 - r)._segment(1), 1 - s))
    )
  ).innerJoin();
}), Nh = l("someCycles", function(t, e) {
  return e._someCyclesBy(0.5, t);
}), Wh = l("often", function(t, e) {
  return e.sometimesBy(0.75, t);
}), $h = l("rarely", function(t, e) {
  return e.sometimesBy(0.25, t);
}), Rh = l("almostNever", function(t, e) {
  return e.sometimesBy(0.1, t);
}), Lh = l("almostAlways", function(t, e) {
  return e.sometimesBy(0.9, t);
}), Ih = l("never", function(t, e) {
  return e;
}), Hh = l("always", function(t, e) {
  return t(e);
});
function _e(t) {
  Array.isArray(t) === !1 && (t = [t]);
  const e = sn();
  return t.every((n) => {
    const s = nn.get(n) ?? n;
    return e[s];
  });
}
const Vh = l("whenKey", function(t, e, n) {
  return n.when(_e(t), e);
}), Fh = l("keyDown", function(t) {
  return t.fmap(_e);
}), H = function(t, e, n = !0) {
  const s = Array.isArray(t), r = Object.keys(t).length;
  return t = Ze(t, d), r === 0 ? v : e.fmap((o) => {
    let c = o;
    return s && (c = n ? Math.round(c) % r : He(Math.round(c), 0, t.length - 1)), t[c];
  });
}, Sa = function(t, e) {
  return Array.isArray(e) && ([e, t] = [t, e]), Ta(t, e);
}, Ta = l("pick", function(t, e) {
  return H(t, e, !1).innerJoin();
}), Ca = l("pickmod", function(t, e) {
  return H(t, e, !0).innerJoin();
}), Dh = l("pickF", function(t, e, n) {
  return n.apply(Sa(t, e));
}), Gh = l("pickmodF", function(t, e, n) {
  return n.apply(Ca(t, e));
}), Qh = l("pickOut", function(t, e) {
  return H(t, e, !1).outerJoin();
}), Uh = l("pickmodOut", function(t, e) {
  return H(t, e, !0).outerJoin();
}), Kh = l("pickRestart", function(t, e) {
  return H(t, e, !1).restartJoin();
}), Xh = l("pickmodRestart", function(t, e) {
  return H(t, e, !0).restartJoin();
}), Yh = l("pickReset", function(t, e) {
  return H(t, e, !1).resetJoin();
}), Zh = l("pickmodReset", function(t, e) {
  return H(t, e, !0).resetJoin();
}), { inhabit: td, pickSqueeze: ed } = l(["inhabit", "pickSqueeze"], function(t, e) {
  return H(t, e, !1).squeezeJoin();
}), { inhabitmod: nd, pickmodSqueeze: sd } = l(["inhabitmod", "pickmodSqueeze"], function(t, e) {
  return H(t, e, !0).squeezeJoin();
}), rd = (t, e) => (e = e.map(d), e.length == 0 ? v : t.fmap((n) => {
  const s = ft(Math.round(n), e.length);
  return e[s];
}).squeezeJoin());
let lt;
try {
  lt = window?.speechSynthesis;
} catch {
  console.warn("cannot use window: not in browser?");
}
let Vt = lt?.getVoices();
function xa(t, e, n) {
  lt.cancel();
  const s = new SpeechSynthesisUtterance(t);
  s.lang = e, Vt = lt.getVoices();
  const r = Vt.filter((o) => o.lang.includes(e));
  typeof n == "number" ? s.voice = r[n % r.length] : typeof n == "string" && (s.voice = r.find((o) => o.name === o)), speechSynthesis.speak(s);
}
const od = l("speak", function(t, e, n) {
  return n.onTrigger((s) => {
    xa(s.value, t, e);
  });
}), Ba = {}, Oa = async (...t) => {
  const e = await Promise.allSettled(t), n = e.filter((s) => s.status === "fulfilled").map((s) => s.value);
  return e.forEach((s, r) => {
    s.status === "rejected" && console.warn(`evalScope: module with index ${r} could not be loaded:`, s.reason);
  }), n.forEach((s) => {
    Object.entries(s).forEach(([r, o]) => {
      globalThis[r] = o, Ba[r] = o;
    });
  }), n;
};
function za(t, e = {}) {
  const { wrapExpression: n = !0, wrapAsync: s = !0 } = e;
  n && (t = `{${t}}`), s && (t = `(async ()=>${t})()`);
  const r = `"use strict";return (${t})`;
  return Function(r)();
}
const Ea = async (t, e, n) => {
  let s = {};
  if (e) {
    const c = e(t, n);
    t = c.output, s = c;
  }
  return { mode: "javascript", pattern: await za(t, { wrapExpression: !!e }), meta: s };
};
class Ma {
  constructor({ onTrigger: e, onToggle: n, getTime: s }) {
    this.started = !1, this.cps = 0.5, this.getTime = s, this.time_at_last_tick_message = 0, this.collator = new tn({ getTargetClockTime: s }), this.onToggle = n, this.latency = 0.1, this.cycle = 0, this.id = Math.round(Date.now() * Math.random()), this.worker = new SharedWorker(new URL(
      /* @vite-ignore */
      "" + new URL("assets/clockworker-ZDiUtESR.js", import.meta.url).href,
      import.meta.url
    )), this.worker.port.start(), this.channel = new BroadcastChannel("strudeltick");
    const r = (c) => {
      const { cps: u, begin: a, end: p, cycle: m, time: y } = c;
      this.cps = u, this.cycle = m;
      const _ = this.collator.calculateOffset(y) + y;
      o(a, p, _), this.time_at_last_tick_message = _;
    }, o = (c, u, a) => {
      if (this.started === !1)
        return;
      this.pattern.queryArc(c, u, { _cps: this.cps, cyclist: "neocyclist" }).forEach((m) => {
        if (m.hasOnset()) {
          const _ = Rt(m.whole.begin - this.cycle, this.cps) + a + this.latency, g = Rt(m.duration, this.cps);
          e?.(m, 0, g, this.cps, _);
        }
      });
    };
    this.channel.onmessage = (c) => {
      if (!this.started)
        return;
      const { payload: u, type: a } = c.data;
      switch (a) {
        case "tick":
          r(u);
      }
    };
  }
  sendMessage(e, n) {
    this.worker.port.postMessage({ type: e, payload: n, id: this.id });
  }
  now() {
    const e = (this.getTime() - this.time_at_last_tick_message) * this.cps;
    return this.cycle + e;
  }
  setCps(e = 1) {
    this.sendMessage("cpschange", { cps: e });
  }
  setCycle(e) {
    this.sendMessage("setcycle", { cycle: e });
  }
  setStarted(e) {
    this.sendMessage("toggle", { started: e }), this.started = e, this.onToggle?.(e);
  }
  start() {
    E("[cyclist] start"), this.setStarted(!0);
  }
  stop() {
    E("[cyclist] stop"), this.collator.reset(), this.setStarted(!1);
  }
  setPattern(e, n = !1) {
    this.pattern = e, n && !this.started && this.start();
  }
  log(e, n, s) {
    const r = s.filter((o) => o.hasOnset());
    console.log(`${e.toFixed(4)} - ${n.toFixed(4)} ${Array(r.length).fill("I").join("")}`);
  }
}
class Pa {
  constructor({
    interval: e,
    onTrigger: n,
    onToggle: s,
    onError: r,
    getTime: o,
    latency: c = 0.1,
    setInterval: u,
    clearInterval: a,
    beforeStart: p
  }) {
    this.started = !1, this.beforeStart = p, this.cps = 0.5, this.num_ticks_since_cps_change = 0, this.lastTick = 0, this.lastBegin = 0, this.lastEnd = 0, this.getTime = o, this.num_cycles_at_cps_change = 0, this.seconds_at_cps_change, this.onToggle = s, this.latency = c, this.clock = ua(
      o,
      // called slightly before each cycle
      (m, y, _, g) => {
        this.num_ticks_since_cps_change === 0 && (this.num_cycles_at_cps_change = this.lastEnd, this.seconds_at_cps_change = m), this.num_ticks_since_cps_change++;
        const q = this.num_ticks_since_cps_change * y * this.cps;
        try {
          const M = this.lastEnd;
          this.lastBegin = M;
          const P = this.num_cycles_at_cps_change + q;
          if (this.lastEnd = P, this.lastTick = m, m < g) {
            console.log("skip query: too late");
            return;
          }
          this.pattern.queryArc(M, P, { _cps: this.cps, cyclist: "cyclist" }).forEach((T) => {
            if (T.hasOnset()) {
              const J = (T.whole.begin - this.num_cycles_at_cps_change) / this.cps + this.seconds_at_cps_change + c, Q = T.duration / this.cps, V = J - m;
              n?.(T, V, Q, this.cps, J), T.value.cps !== void 0 && this.cps != T.value.cps && (this.cps = T.value.cps, this.num_ticks_since_cps_change = 0);
            }
          });
        } catch (M) {
          Gt(M), r?.(M);
        }
      },
      e,
      // duration of each cycle
      0.1,
      0.1,
      u,
      a
    );
  }
  now() {
    if (!this.started)
      return 0;
    const e = this.getTime() - this.lastTick - this.clock.duration;
    return this.lastBegin + e * this.cps;
  }
  setStarted(e) {
    this.started = e, this.onToggle?.(e);
  }
  async start() {
    if (await this.beforeStart?.(), this.num_ticks_since_cps_change = 0, this.num_cycles_at_cps_change = 0, !this.pattern)
      throw new Error("Scheduler: no pattern set! call .setPattern first.");
    E("[cyclist] start"), this.clock.start(), this.setStarted(!0);
  }
  pause() {
    E("[cyclist] pause"), this.clock.pause(), this.setStarted(!1);
  }
  stop() {
    E("[cyclist] stop"), this.clock.stop(), this.lastEnd = 0, this.setStarted(!1);
  }
  async setPattern(e, n = !1) {
    this.pattern = e, n && !this.started && await this.start();
  }
  setCps(e = 0.5) {
    this.cps !== e && (this.cps = e, this.num_ticks_since_cps_change = 0);
  }
  log(e, n, s) {
    const r = s.filter((o) => o.hasOnset());
    console.log(`${e.toFixed(4)} - ${n.toFixed(4)} ${Array(r.length).fill("I").join("")}`);
  }
}
let kt;
function id() {
  if (!kt)
    throw new Error("no time set! use setTime to define a time source");
  return kt();
}
function Ft(t) {
  kt = t;
}
function cd({
  defaultOutput: t,
  onEvalError: e,
  beforeEval: n,
  beforeStart: s,
  afterEval: r,
  getTime: o,
  transpiler: c,
  onToggle: u,
  editPattern: a,
  onUpdateState: p,
  sync: m = !1,
  setInterval: y,
  clearInterval: _,
  id: g,
  mondo: k = !1
}) {
  const q = {
    schedulerError: void 0,
    evalError: void 0,
    code: "// LOADING",
    activeCode: "// LOADING",
    pattern: void 0,
    miniLocations: [],
    widgets: [],
    pending: !1,
    started: !1
  }, M = {
    id: g
  }, P = (b) => {
    Object.assign(q, b), q.isDirty = q.code !== q.activeCode, q.error = q.evalError || q.schedulerError, p?.(q);
  }, Y = {
    onTrigger: Ja({ defaultOutput: t, getTime: o }),
    getTime: o,
    onToggle: (b) => {
      P({ started: b }), u?.(b);
    },
    setInterval: y,
    clearInterval: _,
    beforeStart: s
  }, T = m && typeof SharedWorker < "u" ? new Ma(Y) : new Pa(Y);
  let J = {}, Q = 0, V;
  const Pt = function() {
    return J = {}, Q = 0, V = void 0, v;
  };
  function Jt(b) {
    return b._Pattern ? b.__pure : b;
  }
  const jt = async (b, A = !0) => (b = a?.(b) || b, await T.setPattern(b, A), b);
  Ft(() => T.now());
  const ve = () => T.stop(), ke = () => T.start(), qe = () => T.pause(), Ae = () => T.toggle(), yt = (b) => (T.setCps(Jt(b)), v), Nt = (b) => (T.setCps(Jt(b) / 60), v);
  let Z = [];
  const Se = function(b) {
    return Z.push(b), v;
  }, Te = function(b) {
    return V = b, v;
  }, Ce = () => {
    f.prototype.p = function(A) {
      return typeof A == "string" && (A.startsWith("_") || A.endsWith("_")) ? v : (A === "$" && (A = `$${Q}`, Q++), J[A] = this, this);
    }, f.prototype.q = function(A) {
      return v;
    };
    try {
      for (let A = 1; A < 10; ++A)
        Object.defineProperty(f.prototype, `d${A}`, {
          get() {
            return this.p(A);
          },
          configurable: !0
        }), Object.defineProperty(f.prototype, `p${A}`, {
          get() {
            return this.p(A);
          },
          configurable: !0
        }), f.prototype[`q${A}`] = v;
    } catch (A) {
      console.warn("injectPatternMethods: error:", A);
    }
    const b = l("cpm", function(A, wt) {
      return wt._fast(A / 60 / T.cps);
    });
    return Oa({
      all: Se,
      each: Te,
      hush: Pt,
      cpm: b,
      setCps: yt,
      setcps: yt,
      setCpm: Nt,
      setcpm: Nt
    });
  };
  return { scheduler: T, evaluate: async (b, A = !0, wt = !0) => {
    if (!b)
      throw new Error("no code to evaluate");
    try {
      P({ code: b, pending: !0 }), await Ce(), Ft(() => T.now()), await n?.({ code: b }), Z = [], wt && Pt(), k && (b = `mondolang\`${b}\``);
      let { pattern: O, meta: gt } = await Ea(b, c, M);
      if (Object.keys(J).length) {
        let F = [];
        for (const [bt, xe] of Object.entries(J))
          F.push(xe.withState((Be) => Be.setControls({ id: bt })));
        V && (F = F.map((bt) => V(bt))), O = B(...F);
      } else V && (O = V(O));
      if (Z.length)
        for (let F in Z)
          O = Z[F](O);
      if (!Yt(O)) {
        const F = `got "${typeof evaluated}" instead of pattern`;
        throw new Error(F + (typeof evaluated == "function" ? ", did you forget to call a function?" : "."));
      }
      return E("[eval] code updated"), O = await jt(O, A), P({
        miniLocations: gt?.miniLocations || [],
        widgets: gt?.widgets || [],
        activeCode: b,
        pattern: O,
        evalError: void 0,
        schedulerError: void 0,
        pending: !1
      }), r?.({ code: b, pattern: O, meta: gt }), O;
    } catch (O) {
      E(`[eval] error: ${O.message}`, "error"), console.error(O), P({ evalError: O, pending: !1 }), e?.(O);
    }
  }, start: ke, stop: ve, pause: qe, setCps: yt, setPattern: jt, setCode: (b) => P({ code: b }), toggle: Ae, state: q };
}
const Ja = ({ getTime: t, defaultOutput: e }) => async (n, s, r, o, c) => {
  try {
    (!n.context.onTrigger || !n.context.dominantTrigger) && await e(n, s, r, o, c), n.context.onTrigger && await n.context.onTrigger(n, t(), o, c);
  } catch (u) {
    Gt(u, "getTrigger");
  }
}, ud = function(t, e = {}) {
  const n = document.getElementById("code"), s = "background-image:url(" + t + ");background-size:contain;";
  n.style = s;
  const { className: r } = n, o = (a, p) => {
    ({
      style: () => n.style = s + ";" + p,
      className: () => n.className = p + " " + r
    })[a]();
  }, c = Object.entries(e).filter(([a, p]) => typeof p == "function");
  Object.entries(e).filter(([a, p]) => typeof p == "string").forEach(([a, p]) => o(a, p)), c.length;
}, ad = () => {
  const t = document.getElementById("code");
  t && (t.style = "");
};
E("🌀 @strudel/core loaded 🌀");
globalThis._strudelLoaded && console.warn(
  `@strudel/core was loaded more than once...
This might happen when you have multiple versions of strudel installed. 
Please check with "npm ls @strudel/core".`
);
globalThis._strudelLoaded = !0;
export {
  tn as ClockCollator,
  Pa as Cyclist,
  h as Fraction,
  C as Hap,
  f as Pattern,
  st as State,
  x as TimeSpan,
  we as __chooseWith,
  me as _brandBy,
  _n as _fitslice,
  ye as _irand,
  _e as _keyDown,
  vn as _match,
  ft as _mod,
  ee as _morph,
  qn as _polymeterListSteps,
  Lt as _retime,
  It as _slices,
  Os as accelerate,
  xi as activeLabel,
  ju as ad,
  Al as add,
  Ju as adsr,
  Lh as almostAlways,
  Rh as almostNever,
  Hh as always,
  Ps as amp,
  Us as analyze,
  Zi as anchor,
  Hl as and,
  _p as apply,
  bn as applyN,
  Wu as ar,
  rl as arp,
  sl as arpWith,
  al as arrange,
  ra as as,
  js as att,
  Js as attack,
  Ne as averageArray,
  ud as backgroundImage,
  Ol as band,
  rr as bandf,
  cr as bandq,
  Gs as bank,
  Ye as base64ToUnicode,
  Or as bbexpr,
  Er as bbst,
  Df as beat,
  ar as begin,
  xh as berlin,
  Aa as berlinWith,
  mh as binary,
  ma as binaryN,
  ml as bind,
  ru as binshift,
  Tp as bite,
  oe as bjork,
  Ml as blshift,
  zl as bor,
  ir as bp,
  po as bpa,
  lo as bpattack,
  go as bpd,
  wo as bpdecay,
  oo as bpe,
  ro as bpenv,
  or as bpf,
  ur as bpq,
  Oo as bpr,
  Bo as bprelease,
  Ao as bps,
  qo as bpsustain,
  jp as brak,
  _h as brand,
  bh as brandBy,
  Pl as brshift,
  El as bxor,
  hf as bypass,
  Br as byteBeatExpression,
  zr as byteBeatStartTime,
  el as calculateSteps,
  at as cat,
  Vu as ccn,
  Fu as ccv,
  Ql as ceil,
  Pr as ch,
  Qr as channel,
  Mr as channels,
  wa as choose,
  ga as chooseCycles,
  kh as chooseIn,
  ge as chooseInWith,
  qh as chooseOut,
  Mt as chooseWith,
  Nf as chop,
  Ki as chord,
  Qs as chorus,
  ef as chunk,
  rf as chunkBack,
  ff as chunkBackInto,
  lf as chunkInto,
  of as chunkback,
  pf as chunkbackinto,
  af as chunkinto,
  He as clamp,
  ad as cleanupUi,
  yu as clip,
  wr as coarse,
  Ya as code2hash,
  Eu as color,
  Mu as colour,
  eu as comb,
  Da as compose,
  ep as compress,
  np as compressSpan,
  qc as compressor,
  Tc as compressorAttack,
  Ac as compressorKnee,
  Sc as compressorRatio,
  Cc as compressorRelease,
  sp as compressspan,
  Ga as constant,
  Bn as contract,
  Hu as control,
  Qf as controls,
  rh as cosine,
  oh as cosine2,
  vp as cpm,
  mu as cps,
  ua as createClock,
  Bt as createParam,
  Pu as createParams,
  yr as crush,
  Xr as ctf,
  Du as ctlNum,
  Ei as ctranspose,
  w as curry,
  vu as curve,
  Ur as cut,
  Kr as cutoff,
  Rt as cycleToSeconds,
  Ys as dec,
  Xs as decay,
  zh as degrade,
  Oh as degradeBy,
  Bh as degradeByWith,
  Oi as degree,
  Go as delay,
  Uo as delayfb,
  Qo as delayfeedback,
  Xo as delayspeed,
  ei as delaysync,
  Zo as delayt,
  Yo as delaytime,
  ku as deltaSlide,
  ri as det,
  si as detune,
  Ko as dfb,
  Yi as dict,
  Xi as dictionary,
  _c as dist,
  bc as distort,
  kc as distorttype,
  vc as distortvol,
  Cl as div,
  Do as djf,
  un as drawLine,
  Ar as drive,
  Tn as drop,
  ci as dry,
  Nu as ds,
  ti as dt,
  Sr as duck,
  xr as duckattack,
  Tr as duckdepth,
  Cr as duckonset,
  bu as dur,
  gu as duration,
  Kf as e,
  kp as early,
  Gp as echo,
  Hp as echoWith,
  Vp as echowith,
  nh as eish,
  lr as end,
  Zc as enhance,
  $l as eq,
  Rl as eqt,
  Gt as errorLogger,
  Uf as euclid,
  Zf as euclidLegato,
  th as euclidLegatoRot,
  Yf as euclidRot,
  eh as euclidish,
  Xf as euclidrot,
  Oa as evalScope,
  Ea as evaluate,
  bp as every,
  xn as expand,
  Pc as expression,
  Cn as extend,
  li as fadeInTime,
  ai as fadeOutTime,
  ui as fadeTime,
  Eo as fanchor,
  lp as fast,
  uf as fastChunk,
  rp as fastGap,
  W as fastcat,
  cf as fastchunk,
  op as fastgap,
  Ks as fft,
  gf as filter,
  bf as filterWhen,
  gp as firstOf,
  If as fit,
  D as flatten,
  Gl as floor,
  $s as fm,
  Ls as fmattack,
  Hs as fmdecay,
  Rs as fmenv,
  Ns as fmh,
  Ws as fmi,
  Fs as fmrelease,
  Vs as fmsustain,
  Ds as fmvelocity,
  Is as fmwave,
  ip as focus,
  cp as focusSpan,
  up as focusspan,
  Qa as fractionalArgs,
  cu as frameRate,
  uu as frames,
  Qc as freeze,
  pi as freq,
  Pe as freqToMidi,
  Kl as fromBipolar,
  jc as fshift,
  Nc as fshiftnote,
  Wc as fshiftphase,
  zo as ftype,
  Fl as func,
  Es as gain,
  ct as gap,
  Ai as gat,
  qi as gate,
  se as getControlName,
  sn as getCurrentKeyboardState,
  La as getEventOffsetMs,
  Je as getFreq,
  Fa as getFrequency,
  tl as getPerformanceTimeSeconds,
  Va as getPlayableNoteValue,
  Ha as getSoundIndex,
  id as getTime,
  Ja as getTrigger,
  kf as grow,
  jl as gt,
  Wl as gte,
  Mi as harmonic,
  Za as hash2code,
  ou as hbrick,
  $o as hcutoff,
  sr as hold,
  au as hours,
  Lo as hp,
  ao as hpa,
  uo as hpattack,
  yo as hpd,
  mo as hpdecay,
  so as hpe,
  no as hpenv,
  Ro as hpf,
  Ho as hpq,
  xo as hpr,
  Co as hprelease,
  ko as hps,
  vo as hpsustain,
  Io as hresonance,
  wf as hsl,
  yf as hsla,
  fp as hurry,
  et as id,
  Yc as imag,
  td as inhabit,
  nd as inhabitmod,
  yl as innerBind,
  mp as inside,
  Mp as inv,
  Ep as invert,
  lc as ir,
  vh as irand,
  hc as irbegin,
  pc as iresponse,
  fc as irspeed,
  $n as isControlName,
  qt as isNote,
  $a as isNoteWithOctave,
  Yt as isPattern,
  Ot as isaw,
  ue as isaw2,
  Xp as iter,
  Yp as iterBack,
  Zp as iterback,
  ah as itri,
  lh as itri2,
  Ip as jux,
  Rp as juxBy,
  Lp as juxby,
  Lc as kcutoff,
  kl as keep,
  ql as keepif,
  nn as keyAlias,
  Fh as keyDown,
  Rc as krush,
  Bi as label,
  wp as lastOf,
  wn as late,
  iu as lbrick,
  wu as legato,
  Si as leslie,
  Su as lfo,
  Cp as linger,
  ht as listRange,
  ni as lock,
  Dt as logKey,
  E as logger,
  pr as loop,
  Rf as loopAt,
  Hf as loopAtCps,
  fr as loopBegin,
  dr as loopEnd,
  Lf as loopat,
  Vf as loopatcps,
  hr as loopb,
  mr as loope,
  Zr as lp,
  co as lpa,
  io as lpattack,
  ho as lpd,
  fo as lpdecay,
  eo as lpe,
  to as lpenv,
  Yr as lpf,
  Fo as lpq,
  To as lpr,
  So as lprelease,
  _o as lps,
  bo as lpsustain,
  Ti as lrate,
  Ci as lsize,
  Jl as lt,
  Nl as lte,
  Ut as mapArgs,
  pl as mask,
  Ia as midi2note,
  nt as midiToFreq,
  Zu as midibend,
  $u as midichan,
  Iu as midicmd,
  Ru as midimap,
  Lu as midiport,
  ta as miditouch,
  lu as minutes,
  xl as mod,
  nc as mode,
  Gf as morph,
  dh as mouseX,
  fh as mouseY,
  hh as mousex,
  ph as mousey,
  zi as mtranspose,
  Tl as mul,
  xs as n,
  We as nanFallback,
  Ll as ne,
  Il as net,
  Ih as never,
  jo as noise,
  Bs as note,
  pt as noteToMidi,
  R as nothing,
  Gu as nrpnn,
  Qu as nrpv,
  ji as nudge,
  $ as numeralArgs,
  Ze as objectMap,
  Ni as octave,
  Ji as octaveR,
  ec as octaves,
  Ic as octer,
  Hc as octersub,
  Vc as octersubsub,
  Jp as off,
  tc as offset,
  Wh as often,
  Vl as or,
  Wi as orbit,
  na as oschost,
  sa as oscport,
  wl as outerBind,
  yp as outside,
  $i as overgain,
  Ri as overshape,
  kn as pace,
  Ie as pairs,
  $p as palindrome,
  Li as pan,
  ki as panchor,
  Fi as panorient,
  Ii as panspan,
  Hi as pansplay,
  Vi as panwidth,
  Le as parseFractional,
  Qt as parseNumeral,
  tu as partials,
  hi as patt,
  fi as pattack,
  vi as pcurve,
  mi as pdec,
  di as pdecay,
  _i as penv,
  Ch as perlin,
  ka as perlinWith,
  $r as ph,
  Gr as phasdp,
  Rr as phaser,
  Hr as phasercenter,
  Fr as phaserdepth,
  Wr as phaserrate,
  Lr as phasersweep,
  Vr as phc,
  Dr as phd,
  Ir as phs,
  Sa as pick,
  Dh as pickF,
  Qh as pickOut,
  Yh as pickReset,
  Kh as pickRestart,
  ed as pickSqueeze,
  Ca as pickmod,
  Gh as pickmodF,
  Uh as pickmodOut,
  Zh as pickmodReset,
  Xh as pickmodRestart,
  sd as pickmodSqueeze,
  Re as pipe,
  qu as pitchJump,
  Au as pitchJumpTime,
  ap as ply,
  Kp as plyForEach,
  Up as plyWith,
  cl as pm,
  _l as polyBind,
  ea as polyTouch,
  xt as polymeter,
  ol as polyrhythm,
  Ms as postgain,
  Bl as pow,
  il as pr,
  bi as prel,
  gi as prelease,
  Wp as press,
  Np as pressBy,
  Uu as progNum,
  wi as psus,
  yi as psustain,
  z as pure,
  Jr as pw,
  jr as pwrate,
  Nr as pwsweep,
  L as rand,
  gh as rand2,
  Ah as randcat,
  ya as randrun,
  Xl as range,
  Zl as range2,
  Yl as rangex,
  $h as rarely,
  Di as rate,
  tp as ratio,
  cc as rdim,
  Xc as real,
  Ff as ref,
  l as register,
  i as registerControl,
  d as reify,
  nr as rel,
  er as release,
  it as removeUndefineds,
  tf as repeatCycles,
  Tu as repeatTime,
  cd as repl,
  vf as replicate,
  Vo as resonance,
  gn as rev,
  ac as rfade,
  mf as rib,
  df as ribbon,
  Fc as ring,
  Gc as ringdf,
  Dc as ringf,
  oc as rlp,
  sc as room,
  ic as roomdim,
  uc as roomfade,
  rc as roomlp,
  dc as roomsize,
  $e as rotate,
  Dl as round,
  wc as rsize,
  da as run,
  Rn as s,
  Bf as s_add,
  Sf as s_alt,
  Af as s_cat,
  Mf as s_contract,
  zf as s_expand,
  Ef as s_extend,
  Tf as s_polymeter,
  Of as s_sub,
  Cf as s_taper,
  xf as s_taperlist,
  Pf as s_tour,
  Jf as s_zip,
  mt as saw,
  ce as saw2,
  su as scram,
  wh as scramble,
  oa as scrub,
  pu as seconds,
  Bp as seg,
  xp as segment,
  Qi as semitone,
  yn as seq,
  ll as seqPLoop,
  G as sequence,
  fn as sequenceP,
  vl as set,
  nl as setStringParser,
  Ft as setTime,
  gc as shape,
  zn as shrink,
  On as shrinklist,
  yh as shuffle,
  I as signal,
  v as silence,
  aa as sine,
  ae as sine2,
  mc as size,
  Jn as slice,
  Gi as slide,
  hp as slow,
  sf as slowChunk,
  K as slowcat,
  Zt as slowcatPrime,
  nf as slowchunk,
  nu as smear,
  Ua as sol2note,
  Nh as someCycles,
  jh as someCyclesBy,
  Jh as sometimes,
  Ph as sometimesBy,
  fu as songPtr,
  Ln as sound,
  Ts as source,
  dp as sparsity,
  od as speak,
  ne as speed,
  $f as splice,
  Kt as splitAt,
  ii as spread,
  la as square,
  ih as square2,
  rd as squeeze,
  gl as squeezeBind,
  Oc as squiz,
  Cs as src,
  B as stack,
  ul as stackBy,
  mn as stackCentre,
  hn as stackLeft,
  dn as stackRight,
  sh as steady,
  bl as stepBind,
  An as stepalt,
  N as stepcat,
  jf as steps,
  Pi as stepsPerOctave,
  xc as stretch,
  Wf as striate,
  Xt as stringifyValues,
  fl as struct,
  Ba as strudelScope,
  Qp as stut,
  Fp as stutWith,
  Dp as stutwith,
  Sl as sub,
  hl as superimpose,
  tr as sus,
  Zs as sustain,
  Jc as sustainpedal,
  zp as swing,
  Op as swingBy,
  Ku as sysex,
  Yu as sysexdata,
  Xu as sysexid,
  yc as sz,
  Sn as take,
  le as time,
  Pn as timeCat,
  qf as timecat,
  Ul as toBipolar,
  ze as tokenizeNote,
  En as tour,
  gr as tremolo,
  _r as tremolodepth,
  kr as tremolophase,
  qr as tremoloshape,
  vr as tremoloskew,
  br as tremolosync,
  ch as tri,
  uh as tri2,
  $c as triode,
  Kc as tsdelay,
  hu as uid,
  Mh as undegrade,
  Eh as undegradeBy,
  Xe as unicodeToBase64,
  Ka as uniq,
  Xa as uniqsort,
  Ke as uniqsortr,
  oi as unison,
  Bc as unit,
  Jo as v,
  du as val,
  Ra as valueToMidi,
  zs as velocity,
  Mo as vib,
  No as vibmod,
  Po as vibrato,
  Wo as vmod,
  Ui as voice,
  zc as vowel,
  os as warp,
  us as warpatt,
  cs as warpattack,
  gs as warpdc,
  ls as warpdec,
  as as warpdecay,
  ys as warpdepth,
  As as warpenv,
  _s as warpmode,
  ms as warprate,
  ds as warprel,
  hs as warprelease,
  ws as warpshape,
  bs as warpskew,
  fs as warpsus,
  ps as warpsustain,
  Ss as warpsync,
  Ec as waveloss,
  qs as wavetablePhaseRand,
  Hn as wavetablePosition,
  is as wavetableWarp,
  vs as wavetableWarpMode,
  Sh as wchoose,
  _a as wchooseCycles,
  Pp as when,
  Vh as whenKey,
  dl as withValue,
  _f as within,
  Th as wrandcat,
  In as wt,
  Dn as wtatt,
  Fn as wtattack,
  ss as wtdc,
  Qn as wtdec,
  Gn as wtdecay,
  es as wtdepth,
  Vn as wtenv,
  ks as wtphaserand,
  Zn as wtrate,
  Yn as wtrel,
  Xn as wtrelease,
  ns as wtshape,
  rs as wtskew,
  Kn as wtsus,
  Un as wtsustain,
  ts as wtsync,
  jn as xfade,
  Uc as xsdelay,
  Bu as zcrush,
  Ou as zdelay,
  Mn as zip,
  At as zipWith,
  xu as zmod,
  Cu as znoise,
  qp as zoom,
  Ap as zoomArc,
  Sp as zoomarc,
  _u as zrand,
  zu as zzfx
};
